<!DOCTYPE html><html lang="en" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>BItest2 - 尝试解读两类文献思路CIBERSOFT和差异分析 | PiguDog的小屋</title><meta name="keywords" content="BItest"><meta name="author" content="pigudog,2205638116@qq.com"><meta name="copyright" content="pigudog"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="Quetion解读这两个思路的文献第一个是 CIBERSORT 思路： 根据文章：Cancer Med. 2018 Sep; 题目是：Profiles of immune infiltration in colorectal cancer and their clinical significant: A gene expression‐based study 研究者也是使用 CIBERSORT">
<meta property="og:type" content="article">
<meta property="og:title" content="BItest2 - 尝试解读两类文献思路CIBERSOFT和差异分析">
<meta property="og:url" content="http://example.com/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/index.html">
<meta property="og:site_name" content="PiguDog的小屋">
<meta property="og:description" content="Quetion解读这两个思路的文献第一个是 CIBERSORT 思路： 根据文章：Cancer Med. 2018 Sep; 题目是：Profiles of immune infiltration in colorectal cancer and their clinical significant: A gene expression‐based study 研究者也是使用 CIBERSORT">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png">
<meta property="article:published_time" content="2022-11-16T16:00:00.000Z">
<meta property="article:modified_time" content="2022-11-17T10:02:20.642Z">
<meta property="article:author" content="pigudog">
<meta property="article:tag" content="BItest">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="http://example.com/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: {"defaultEncoding":1,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: 'days',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: {"limitCount":50,"languages":{"author":"Author: pigudog","link":"Link: ","source":"Source: PiguDog的小屋","info":"Copyright is owned by the author. For commercial reprints, please contact the author for authorization. For non-commercial reprints, please indicate the source."}},
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'BItest2 - 尝试解读两类文献思路CIBERSOFT和差异分析',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2022-11-17 18:02:20'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.2.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/img/avatar.gif" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">70</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">12</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">16</div></a></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> Movie</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">PiguDog的小屋</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> Movie</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">BItest2 - 尝试解读两类文献思路CIBERSOFT和差异分析<a class="post-edit-link" href="https://github.com/pigudog/_posts/生信学徒/BItest2.md" title="Edited on" target="_blank"><i class="fas fa-pencil-alt"></i></a></h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2022-11-16T16:00:00.000Z" title="Created 2022-11-17 00:00:00">2022-11-17</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2022-11-17T10:02:20.642Z" title="Updated 2022-11-17 18:02:20">2022-11-17</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/medical/">medical</a><i class="fas fa-angle-right post-meta-separator"></i><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/medical/BItest/">BItest</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="BItest2 - 尝试解读两类文献思路CIBERSOFT和差异分析"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="Quetion"><a href="#Quetion" class="headerlink" title="Quetion"></a>Quetion</h1><p><strong>解读这两个思路的文献</strong><br>第一个是 CIBERSORT 思路：</p>
<p>根据文章：<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6144159/#">Cancer Med</a>. 2018 Sep; 题目是：Profiles of immune infiltration in colorectal cancer and their clinical significant: A gene expression‐based study 研究者也是使用 <strong>CIBERSORT</strong> 在 colorectal cancer (CRC) 数据里面对22种免疫细胞计算得分，从GEO拿到1700数据，TCGA近600，</p>
<p>根据文章：<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/27959923#">PLoS Med.</a> 2016 Dec 13；题目是：Patterns of Immune Infiltration in Breast Cancer and Their Clinical Implications: A Gene-Expression-Based Retrospective Study. 研究者也是使用<strong>CIBERSORT</strong>在11,000 tumours 看 22 subsets of immune cells的比例。</p>
<p>根据文章 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/31351430#">Biomed Pharmacother.</a> 2019 Jul ； 题目是 ： Profiles of immune cell infiltration and immune-related genes in the tumor microenvironment of colorectal cancer.  研究者也是使用 <strong>CIBERSORT</strong>, a deconvolution algorithm, was used to analyze the infiltration of 22 immune cell types in the tumor microenvironment and immune-related gene expression in 404 CRC and 40 adjacent non-tumorous tissues. 使用的是纯粹的TCGA数据：HTSeq-FPKM</p>
<p>第二个是纯粹的数据集差异分析思路，<strong>差异分析，数据库富集，</strong> <strong>string</strong> <strong>数据库获取</strong> <strong>ppi</strong> <strong>，</strong> <strong>hub</strong> <strong>基因策略</strong>一直都有效<br><a target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/31895772/">https://pubmed.ncbi.nlm.nih.gov/31895772/</a><br><a target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/30140156/">https://pubmed.ncbi.nlm.nih.gov/30140156/</a><br><a target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/31076281/">https://pubmed.ncbi.nlm.nih.gov/31076281/</a><br><a target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/29081411/">https://pubmed.ncbi.nlm.nih.gov/29081411/</a><br><a target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/32358373/">https://pubmed.ncbi.nlm.nih.gov/32358373/</a></p>
<h1 id="Test1-for-CIBERSOFT"><a href="#Test1-for-CIBERSOFT" class="headerlink" title="Test1 - for CIBERSOFT"></a>Test1 - for CIBERSOFT</h1><h2 id="肿瘤免疫微环境和肿瘤免疫浸润"><a href="#肿瘤免疫微环境和肿瘤免疫浸润" class="headerlink" title="肿瘤免疫微环境和肿瘤免疫浸润"></a>肿瘤免疫微环境和肿瘤免疫浸润</h2><ul>
<li>2018-2019的生信分析热点–肿瘤免疫微环境和肿瘤免疫浸润</li>
<li><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/298355310">ref1</a></li>
<li><a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1622907">ref2</a></li>
<li>那么什么是肿瘤免疫微环境和肿瘤免疫浸入？<ul>
<li>首先我们要明白肿瘤组织不只是包含肿瘤细胞，同时包含部分基质细胞和免疫细胞，这个就是肿瘤的免疫微环境，而肿瘤对应的占比就是肿瘤纯度tumor purity。这些免疫细胞之间，以及肿瘤细胞和免疫细胞之间存在着千丝万缕，免疫细胞多种多样，所以所谓免疫微环境，或者说免疫浸润的分析，本质上，就是搞清楚肿瘤组织当中免疫细胞的构成比例。</li>
<li>免疫治疗指的是通过激活人体本身的免疫系统，依靠自身的免疫机制消灭癌细胞。目前已经在多种肿瘤如黑色素瘤、非小细胞肺癌、肾癌和前列腺等实体瘤的治疗中展现出强大的抗肿瘤活性，多个肿瘤免疫治疗药物已经获得美国FDA批准临床应用。</li>
<li><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/Pasted%20image%2020221116195735.png"></li>
<li>与此同时，免疫细胞里面又可以分为多种类：<ul>
<li>B细胞</li>
<li>NK细胞</li>
<li>T细胞</li>
<li>DC细胞等</li>
</ul>
</li>
<li>那么如何判断免疫细胞的构成呢？<ul>
<li>一类是“实测法”：高精度的单细胞测序，Single cell RNA-seq。为了搞清楚构成，索性一个一个细胞去分开测序，把细胞表达都测了，通过一些标记基因，来判断到底这里面有多少类型的免疫细胞，以及每种细胞的构成比例。</li>
<li>另一种方法，那就是“推测法”，通过 bulk RNA-seq进行推测。我们常规的普通RNA测序和芯片结果就是把所有细胞统一裂解抽提RNA，那么检测到的基因表达中，就是各种免疫细胞和肿瘤细胞混杂在一起的表达，通过一些算法，我们可以从这个混杂的表达谱中推断免疫细胞的构成比，常用的软件有CIBETSORT、TIMER、EPIC等等。</li>
</ul>
</li>
<li>那儿么这些细胞通常会表达一些特定的基因，那么运用mRNA的表达矩阵，通过一系列算法(ssGSEA, Cibersoft, Timer, EPIC)可以去计算不同种类的免疫细胞在不同病人中的免疫浸润程度<ul>
<li><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/Pasted%20image%2020221116201211.png" alt="400"></li>
</ul>
</li>
</ul>
</li>
<li>那么这些免疫浸润算法得到的数值结果可以作为表型数据与临床数据以及其他基因表型数据结合进行进一步的分析</li>
<li>分析工具简介<ul>
<li>ESTIMATE 算法，肿瘤纯度，免疫细胞及基质细胞评分</li>
<li>TIMER: 免疫细胞浸润，把别人算法也纳入进自己的网站了，也就是说用TIMER可以替代cibersoft EPIC等工具。</li>
<li>ssGSEA: 老牌算法，简单易用</li>
<li><strong>Cibersoft：常规免疫浸润算法，使用较多，被TIMER承包了，Cibersoft的图用R绘制是没问题的</strong></li>
<li>EPIC：最简单的工具</li>
<li>xcell：可以分析样本的细胞组成，可惜我没用过</li>
<li>quantiseq：肿瘤免疫细胞定量分析，又被TIMER承包了</li>
</ul>
</li>
</ul>
<h2 id="CIBRSOFT"><a href="#CIBRSOFT" class="headerlink" title="CIBRSOFT"></a>CIBRSOFT</h2><ul>
<li><a target="_blank" rel="noopener" href="https://www.nature.com/articles/nmeth.3337">ref1</a></li>
<li><a target="_blank" rel="noopener" href="https://blog.csdn.net/u013429737/article/details/116044812">ref2</a><br>CIBERSORT is an <strong>analytical tool</strong> from the Alizadeh Lab developed by Newman et al. to provide an estimation of the abundances of member cell types in a mixed cell population, using gene expression data.</li>
<li>输入基因表达矩阵，输出样本中的各种细胞类型的丰度。<br>CIBERSORTx, the <em>next</em> generation version of CIBERSORT, is now available (Newman et al.), with support for single-cell RNA-seq and cell type-specific gene expression purification. We recommend moving over to the CIBERSORTx website. For users who registered with CIBERSORT prior to 2018, you may log in with your CIBERSORT account credentials, otherwise please register for a new account.</li>
</ul>
<p>CIBERSORT ：<a target="_blank" rel="noopener" href="https://cibersort.stanford.edu/">旧版官网入口</a><br>现在新版本名字加了一个x，叫CIBERSORTx，主要增加了适配大量组织(bulk tissue)样本单细胞测序方面的功能，<a target="_blank" rel="noopener" href="https://cibersortx.stanford.edu/">新版入口</a><br>CIBERSORT是2015年发表于Nature Methods上的文章（IF：47.990）： <a target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/25822800/">https://pubmed.ncbi.nlm.nih.gov/25822800/</a></p>
<h3 id="什么是CIBERSOFT"><a href="#什么是CIBERSOFT" class="headerlink" title="什么是CIBERSOFT"></a>什么是CIBERSOFT</h3><p><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/20210423103043178.png"></p>
<ul>
<li>CIBERSORT的设计之初的核心目标是“predicting fractions of multiple cell types in gene expression profiles (GEPs)”，可能例子里22个免疫细胞 signature让人太过印象深刻，几乎全互联网的教程都是基于免疫细胞的，CIBERSORT也被认为是专门用来量化免疫浸润的</li>
</ul>
<h3 id="HOW-TO-USE-CIBERSOFT"><a href="#HOW-TO-USE-CIBERSOFT" class="headerlink" title="HOW TO USE CIBERSOFT"></a>HOW TO USE CIBERSOFT</h3><ul>
<li>这里只做简单的操作实践</li>
<li>例子数据来自：<a target="_blank" rel="noopener" href="https://github.com/zomithex/CIBERSORT">https://github.com/zomithex/CIBERSORT</a>，当然也可以从官网下，就是LM22后缀不一样，一个是txt,另一个是csv。</li>
<li>CIBERSORT可以在线用（查看官网），也可以在本地用R环境跑</li>
<li>CIBERSORT是基于线性支持向量回归(SVR)开发的，gene作为输入在计算过程中去卷积化。</li>
<li>CIBERSORT主要需要v-SVR算法的库， 用R去执行CIBERSORT的话，需要加载 R package e1071作为前置条件。<br>下面就是进行分析了，在进行分析的时候我们需要准备三个文件：</li>
</ul>
<ol>
<li>LM22.txt（可以从CIBERSORT网站下载，这个就是22种免疫细胞的参考marker基因表达）</li>
<li>CIBERSORT.R（CIBERSORT源代码，从官网下载）</li>
<li>expression.txt（基因表达谱文件，就是上面说的肺腺癌表达谱）<br>然后，运行CIBERSORT非常简单，三个文件放到一个文件夹中，在RStudio中设置工作路径，运行如下代码即可：<br>简介：<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">source<span class="punctuation">(</span><span class="string">&quot;CIBERSORT.R&quot;</span><span class="punctuation">)</span> </span><br><span class="line"></span><br><span class="line"><span class="comment"># Define LM22 file </span></span><br><span class="line">LM22.file <span class="operator">&lt;-</span> <span class="string">&quot;LM22.txt&quot;</span> </span><br><span class="line">exp.file <span class="operator">&lt;-</span> <span class="string">&quot;datExp.processed.txt&quot;</span> </span><br><span class="line">TME.results <span class="operator">=</span> CIBERSORT<span class="punctuation">(</span>LM22.file<span class="punctuation">,</span> exp.file<span class="punctuation">,</span> perm <span class="operator">=</span> <span class="number">1000</span><span class="punctuation">,</span> QN <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span> </span><br><span class="line"><span class="comment"># output CIBERSORT results </span></span><br><span class="line">write.table<span class="punctuation">(</span>TME.results<span class="punctuation">,</span> <span class="string">&quot;TME.results.output.txt&quot;</span><span class="punctuation">,</span> sep <span class="operator">=</span> <span class="string">&quot;\t&quot;</span><span class="punctuation">,</span> row.names <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span> col.names <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span> <span class="built_in">quote</span> <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/37562ufj2g.png"></li>
</ol>
<ul>
<li>每一行一个样本，每一列一种细胞，总共22中细胞</li>
<li>数值即免疫细胞的所占的比例</li>
</ul>
<p>我们自己尝试一下</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># TCGA数据库大肠癌的100个样本</span></span><br><span class="line">library<span class="punctuation">(</span><span class="string">&quot;limma&quot;</span><span class="punctuation">)</span>  </span><br><span class="line">library<span class="punctuation">(</span><span class="string">&#x27;e1071&#x27;</span><span class="punctuation">)</span></span><br><span class="line">source<span class="punctuation">(</span><span class="string">&quot;CIBERSORT.R&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="built_in">exp</span> <span class="operator">&lt;-</span> read.csv<span class="punctuation">(</span><span class="string">&quot;TCGA_ROAD_100.csv&quot;</span><span class="punctuation">,</span>header<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 整理一下行名，列名，删除表达特别低的基因</span></span><br><span class="line"><span class="built_in">exp</span><span class="operator">=</span>as.matrix<span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">)</span></span><br><span class="line">rownames<span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">)</span><span class="operator">=</span><span class="built_in">exp</span><span class="punctuation">[</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">]</span></span><br><span class="line"><span class="built_in">exp</span><span class="operator">=</span><span class="built_in">exp</span><span class="punctuation">[</span><span class="punctuation">,</span><span class="number">2</span><span class="operator">:</span>ncol<span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"><span class="built_in">dimnames</span><span class="operator">=</span><span class="built_in">list</span><span class="punctuation">(</span>rownames<span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">)</span><span class="punctuation">,</span>colnames<span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">data<span class="operator">=</span>matrix<span class="punctuation">(</span><span class="built_in">as.numeric</span><span class="punctuation">(</span>as.matrix<span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span>nrow<span class="operator">=</span>nrow<span class="punctuation">(</span><span class="built_in">exp</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="built_in">dimnames</span><span class="operator">=</span><span class="built_in">dimnames</span><span class="punctuation">)</span></span><br><span class="line">data<span class="operator">=</span>avereps<span class="punctuation">(</span>data<span class="punctuation">)</span></span><br><span class="line">data<span class="operator">=</span>data<span class="punctuation">[</span>rowMeans<span class="punctuation">(</span>data<span class="punctuation">)</span><span class="operator">&gt;</span><span class="number">0</span><span class="punctuation">,</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#把准备输入CIBERSORT的数据保存一下</span></span><br><span class="line">v <span class="operator">&lt;-</span>voom<span class="punctuation">(</span>data<span class="punctuation">,</span> plot <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span> save.plot <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line">out<span class="operator">=</span>v<span class="operator">$</span>E</span><br><span class="line">out<span class="operator">=</span>rbind<span class="punctuation">(</span>ID<span class="operator">=</span>colnames<span class="punctuation">(</span>out<span class="punctuation">)</span><span class="punctuation">,</span>out<span class="punctuation">)</span></span><br><span class="line">write.table<span class="punctuation">(</span>out<span class="punctuation">,</span>file<span class="operator">=</span><span class="string">&quot;TCGA_100_ready.txt&quot;</span><span class="punctuation">,</span>sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">,</span><span class="built_in">quote</span><span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span>col.names<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>           </span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">results<span class="operator">=</span>CIBERSORT<span class="punctuation">(</span><span class="string">&quot;LM22.txt&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TCGA_100_ready.txt&quot;</span><span class="punctuation">,</span> perm<span class="operator">=</span><span class="number">100</span><span class="punctuation">,</span> QN<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">write.csv<span class="punctuation">(</span>results<span class="punctuation">,</span><span class="string">&quot;TCGA_100_CIBERSORT_Output.csv&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>然后就可以进行数据可视化</p>
<h3 id="Cibersoft-Data-visualization"><a href="#Cibersoft-Data-visualization" class="headerlink" title="Cibersoft Data visualization"></a>Cibersoft Data visualization</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">pkgs <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;matrixStats&quot;</span><span class="punctuation">,</span> <span class="string">&quot;pheatmap&quot;</span><span class="punctuation">,</span> <span class="string">&quot;RColorBrewer&quot;</span><span class="punctuation">,</span> <span class="string">&quot;tidyverse&quot;</span><span class="punctuation">,</span> <span class="string">&quot;cowplot&quot;</span><span class="punctuation">,</span><span class="string">&quot;ggpubr&quot;</span><span class="punctuation">,</span><span class="string">&quot;bslib&quot;</span><span class="punctuation">,</span><span class="string">&quot;ggthemes&quot;</span><span class="punctuation">)</span></span><br><span class="line">lapply<span class="punctuation">(</span>pkgs<span class="punctuation">,</span> library<span class="punctuation">,</span> character.only <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># Read in results ---------------------------------------------------------</span></span><br><span class="line"></span><br><span class="line">cibersort_raw <span class="operator">&lt;-</span> read.csv<span class="punctuation">(</span><span class="string">&quot;TCGA_100_CIBERSORT_Output.csv&quot;</span><span class="punctuation">,</span>row.names <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span>header <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>dplyr<span class="punctuation">)</span></span><br><span class="line">library<span class="punctuation">(</span>tidyr<span class="punctuation">)</span></span><br><span class="line">dd1 <span class="operator">&lt;-</span> cibersort_raw <span class="operator">%&gt;%</span></span><br><span class="line">  as.data.frame<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">  rownames_to_column<span class="punctuation">(</span><span class="string">&quot;sample&quot;</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">  pivot_longer<span class="punctuation">(</span>cols <span class="operator">=</span> <span class="number">2</span><span class="operator">:</span><span class="number">23</span><span class="punctuation">,</span></span><br><span class="line">               names_to <span class="operator">=</span> <span class="string">&quot;CellType&quot;</span><span class="punctuation">,</span></span><br><span class="line">               values_to <span class="operator">=</span> <span class="string">&quot;Composition&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">plot.info <span class="operator">&lt;-</span> dd1<span class="punctuation">[</span><span class="punctuation">,</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">5</span><span class="punctuation">,</span><span class="number">1</span><span class="punctuation">,</span><span class="number">6</span><span class="punctuation">)</span><span class="punctuation">]</span>          </span><br></pre></td></tr></table></figure>

<p>首先，我们用箱线图表示22种免疫细胞在所有样本种的比例</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">ggboxplot<span class="punctuation">(</span></span><br><span class="line">  plot.info<span class="punctuation">,</span></span><br><span class="line">  x <span class="operator">=</span> <span class="string">&quot;CellType&quot;</span><span class="punctuation">,</span></span><br><span class="line">  y <span class="operator">=</span> <span class="string">&quot;Composition&quot;</span><span class="punctuation">,</span></span><br><span class="line">  color <span class="operator">=</span> <span class="string">&quot;black&quot;</span><span class="punctuation">,</span></span><br><span class="line">  fill <span class="operator">=</span> <span class="string">&quot;CellType&quot;</span><span class="punctuation">,</span></span><br><span class="line">  xlab <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">,</span></span><br><span class="line">  ylab <span class="operator">=</span> <span class="string">&quot;Cell composition&quot;</span><span class="punctuation">,</span></span><br><span class="line">  main <span class="operator">=</span> <span class="string">&quot;TME Cell composition&quot;</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  theme_base<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  theme<span class="punctuation">(</span>axis.text.x <span class="operator">=</span> element_text<span class="punctuation">(</span></span><br><span class="line">    angle <span class="operator">=</span> <span class="number">90</span><span class="punctuation">,</span></span><br><span class="line">    hjust <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">    vjust <span class="operator">=</span> <span class="number">1</span></span><br><span class="line">  <span class="punctuation">)</span><span class="punctuation">)</span>          </span><br></pre></td></tr></table></figure>
<p><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/20210423180710642.png"></p>
<p>为了比较不同分期之间的差异，我们还需要读入一个样本分期的表型文件</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># box plot group by GRADE </span></span><br><span class="line">ggboxplot<span class="punctuation">(</span> </span><br><span class="line">	plot.info<span class="punctuation">,</span> </span><br><span class="line">	x <span class="operator">=</span> <span class="string">&quot;CellType&quot;</span><span class="punctuation">,</span> </span><br><span class="line">	y <span class="operator">=</span> <span class="string">&quot;Composition&quot;</span><span class="punctuation">,</span> </span><br><span class="line">	color <span class="operator">=</span> <span class="string">&quot;black&quot;</span><span class="punctuation">,</span> </span><br><span class="line">	fill <span class="operator">=</span> <span class="string">&quot;GRADE&quot;</span><span class="punctuation">,</span> </span><br><span class="line">	xlab <span class="operator">=</span> <span class="string">&quot;&quot;</span><span class="punctuation">,</span> </span><br><span class="line">	ylab <span class="operator">=</span> <span class="string">&quot;Cell composition&quot;</span><span class="punctuation">,</span> </span><br><span class="line">	main <span class="operator">=</span> <span class="string">&quot;TME Cell composition group by GRADE&quot;</span> </span><br><span class="line"><span class="punctuation">)</span> <span class="operator">+</span> </span><br><span class="line">	stat_compare_means<span class="punctuation">(</span> </span><br><span class="line">	label <span class="operator">=</span> <span class="string">&quot;p.signif&quot;</span><span class="punctuation">,</span> 														</span><br><span class="line">	method <span class="operator">=</span> <span class="string">&quot;t.test&quot;</span><span class="punctuation">,</span> 														</span><br><span class="line">	ref.group <span class="operator">=</span> <span class="string">&quot;.all.&quot;</span><span class="punctuation">,</span> 														</span><br><span class="line">	hide.ns <span class="operator">=</span> <span class="built_in">T</span> </span><br><span class="line">	<span class="punctuation">)</span> <span class="operator">+</span> 														</span><br><span class="line">	theme_base<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span> 														</span><br><span class="line">	theme<span class="punctuation">(</span>axis.text.x <span class="operator">=</span> element_text<span class="punctuation">(</span> 														</span><br><span class="line">	angle <span class="operator">=</span> <span class="number">90</span><span class="punctuation">,</span> 													</span><br><span class="line">	hjust <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span> 															</span><br><span class="line">	vjust <span class="operator">=</span> <span class="number">1</span> 															<span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<p><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/Pasted%20image%2020221116205526.png"></p>
<p>为了更加直观地感受不同细胞在各个样本中的比例差异，在这里非常适合做一个堆积的比例柱状图</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">ggbarplot<span class="punctuation">(</span></span><br><span class="line">  plot.info<span class="punctuation">,</span></span><br><span class="line">  x <span class="operator">=</span> <span class="string">&quot;sample&quot;</span><span class="punctuation">,</span></span><br><span class="line">  y <span class="operator">=</span> <span class="string">&quot;Composition&quot;</span><span class="punctuation">,</span></span><br><span class="line">  size <span class="operator">=</span> <span class="number">0</span><span class="punctuation">,</span></span><br><span class="line">  fill <span class="operator">=</span> <span class="string">&quot;CellType&quot;</span><span class="punctuation">,</span></span><br><span class="line">  color <span class="operator">=</span> <span class="string">&quot;CellType&quot;</span><span class="punctuation">,</span></span><br><span class="line"></span><br><span class="line"><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  theme_base<span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">+</span></span><br><span class="line">  theme<span class="punctuation">(</span></span><br><span class="line">    axis.text.x <span class="operator">=</span> element_text<span class="punctuation">(</span></span><br><span class="line">      angle <span class="operator">=</span> <span class="number">90</span><span class="punctuation">,</span></span><br><span class="line">      hjust <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">      vjust <span class="operator">=</span> <span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">      size <span class="operator">=</span> <span class="number">1</span></span><br><span class="line">    <span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    legend.position <span class="operator">=</span> <span class="string">&quot;bottom&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/20210423181010423.png"></p>
<h2 id="001-Profiles-of-immune-infiltration-in-colorectal-cancer-and-their-clinical-significant-A-gene-expression‐based-study"><a href="#001-Profiles-of-immune-infiltration-in-colorectal-cancer-and-their-clinical-significant-A-gene-expression‐based-study" class="headerlink" title="001 Profiles of immune infiltration in colorectal cancer and their clinical significant: A gene expression‐based study"></a>001 Profiles of immune infiltration in colorectal cancer and their clinical significant: A gene expression‐based study</h2><p>本地：<a href="BItest2/BItest2_1.pdf">BItest2_1</a><br>期刊： Cancer Med; 3, 4.711, 2018</p>
<p>结直肠癌免疫浸润的概况及其临床意义:一项基于基因表达的研究</p>
<h3 id="Available-of-data-and-materials"><a href="#Available-of-data-and-materials" class="headerlink" title="Available of data and materials"></a>Available of data and materials</h3><p>The datasets supporting the conclusions of this article are available in the in Gene Expression Omnibus (GEO) with the accession numbers listed as below: <strong>GSE12945, GSE14333, GSE17536, GSE17537, GSE24551, GSE29623, GSE37892, GSE39582, GSE42258</strong>. </p>
<h3 id="Conclusion"><a href="#Conclusion" class="headerlink" title="Conclusion"></a>Conclusion</h3><ul>
<li>As a result, profiles of immune infiltration vary significantly between <strong>paired cancer and paracancerous tissue</strong> and the variation could characterize the individual differences. <ul>
<li>结果表明，成对癌组织和癌旁组织的免疫浸润谱差异显著，这种差异可以表征个体差异。</li>
</ul>
</li>
<li>Of the cell <strong>subpopulations</strong> investigated, tumors <strong>lacking M1 macrophages</strong> or with an <strong>increased</strong> number of <strong>M2 macrophages,</strong> <strong>eosinophils</strong>, and <strong>neutrophils</strong> were associated with the poor prognosis. Unsupervised clustering analysis using immune cell proportions revealed five subgroups of tumors, largely defined by the balance between macrophages M1, M2, and NK resting cells, with distinct survival patterns, and associated with well- established molecular subtype<ul>
<li>在所调查的细胞亚群中，缺乏M1巨噬细胞或M2巨噬细胞、嗜酸性粒细胞和中性粒细胞数量增加的肿瘤与不良预后相关。使用免疫细胞比例进行无监督聚类分析发现了5个肿瘤亚组，主要由巨噬细胞M1、M2和NK静息细胞之间的平衡确定，具有不同的生存模式，并与已确定的分子亚型相关</li>
</ul>
</li>
<li>Collectively, our data suggest that <strong>subtle differences</strong> in the cellular composition of the immune infiltrate in CRC appear to exist, and these differences <strong>are likely to be</strong> important determinants of both prognosis and response to treatment.<ul>
<li>CRC中免疫浸润的细胞组成的微妙的差异，可能是预后和治疗反应的重要决定因素。</li>
</ul>
</li>
</ul>
<h3 id="Noun-explanation"><a href="#Noun-explanation" class="headerlink" title="Noun explanation"></a>Noun explanation</h3><ul>
<li>Immune infiltration of colorectal cancer (CRC)<ul>
<li>结直肠癌的免疫浸润</li>
<li>Immune infiltration</li>
</ul>
</li>
<li>deconvolution algorithm (known as CIBERSORT) <ul>
<li>反褶积算法(称为CIBERSORT)</li>
</ul>
</li>
<li>Tumor- infiltrating immune cells (TIICs)<ul>
<li>肿瘤浸润性免疫细胞</li>
</ul>
</li>
<li></li>
</ul>
<h3 id="Analysis-of-Process"><a href="#Analysis-of-Process" class="headerlink" title="Analysis of Process"></a>Analysis of Process</h3><ul>
<li>Data acquistion: <ul>
<li>from GEO and TCGA</li>
<li>small sample size (N &lt; 50) and duplications were excluded</li>
<li>Preprocessing and aggregation of raw data were performed according to the <strong>robust multi- array average algorithm</strong></li>
<li>for TCGA datasets, RNA sequencing data were transformed using voom27(variance modeling at the observational level) <ul>
<li>count data -&gt; more similar to data from microarrays</li>
</ul>
</li>
<li><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/Pasted%20image%2020221116210659.png"></li>
</ul>
</li>
<li>Evaluation of tumor- infiltrating immune cells<ul>
<li>Normalized gene expression data were used to infer the relative proportions of 22 types of infiltrating immune cells using the <strong>CIBERSORT algorithm</strong> as et al previously reported.</li>
</ul>
</li>
<li>Immunohistochemical detection of immune cell types</li>
<li>Data Analysis - survival analysis<ul>
<li>Cases with a CIBERSORT P- value of &lt;0.05 were included in the main <strong>survival analysis.</strong></li>
<li>Associations between inferred proportions of immune cell types and survival were tested using <strong>Cox regression.</strong></li>
<li>The associations of immune cell infiltrate and corresponding disease- free survival were analyzed by <strong>Kaplan- Meier curves</strong> and evaluated using <strong>log- rank test</strong></li>
<li><strong>Hierarchical clustering</strong> of these data by <strong>Ward</strong>’s method was conducted across all samples.<ul>
<li>a combination of the <em>elbow method</em> and the gap statistic</li>
</ul>
</li>
<li></li>
</ul>
</li>
</ul>
<h2 id="002-Patterns-of-Immune-Infiltration-in-Breast-Cancer-and-Their-Clinical-Implications-A-Gene-Expression-Based-Retrospective-Study"><a href="#002-Patterns-of-Immune-Infiltration-in-Breast-Cancer-and-Their-Clinical-Implications-A-Gene-Expression-Based-Retrospective-Study" class="headerlink" title="002 Patterns of Immune Infiltration in Breast Cancer and Their Clinical Implications: A Gene-Expression-Based Retrospective Study"></a>002 Patterns of Immune Infiltration in Breast Cancer and Their Clinical Implications: A Gene-Expression-Based Retrospective Study</h2><p>本地：<a href="BItest2/BItest2_2.pdf">BItest2_2</a><br>期刊：PLoS Med, 1, 11.613, 2016</p>
<p>乳腺癌免疫浸润模式及其临床意义:一项基于基因表达的回顾性研究</p>
<h3 id="Conclustion"><a href="#Conclustion" class="headerlink" title="Conclustion"></a>Conclustion</h3><p>Large differences in the cellular composition of the immune infiltrate in breast tumours appear to exist, and these differences are likely to be important determinants of both prognosis and response to treatment. In particular, <strong>macrophages</strong> emerge as a possible target for novel therapies. Detailed analysis of the cellular immune response in tumours has the potential to enhance clinical prediction and to identify candidates for immunotherapy.</p>
<h3 id="available-of-data"><a href="#available-of-data" class="headerlink" title="available of data"></a>available of data</h3><p>文中的一个链接似乎打不开：</p>
<ul>
<li><a target="_blank" rel="noopener" href="http://compbio.dfci.harvard.edu/pubs/sbtpaper/">http://compbio.dfci.harvard.edu/pubs/sbtpaper/</a><br>可以在文末找到table的链接<br><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/Pasted%20image%2020221117170509.png"></li>
<li>放在supporting information里面，包括了没有放在文中的图片以及各类的表格</li>
</ul>
<h3 id="Analysis-of-process"><a href="#Analysis-of-process" class="headerlink" title="Analysis of process"></a>Analysis of process</h3><p>method的段进行分析</p>
<ul>
<li>Gene Expression Datasets<ul>
<li>Samples with <strong>Pearson correlations</strong> of 0.99 or greater were considered replicates (n &#x3D; 248).</li>
</ul>
</li>
<li>Molecular Subtyping and Inference of Infiltrating Immune Immune Cells<ul>
<li>Normalised gene expression data were used to infer the relative proportions of 22 types of infiltrating immune cells using the <strong>CIBERSORT algorithm.</strong></li>
<li>For the TCGA dataset [3], RNA sequencing data were transformed using <strong>voom</strong></li>
</ul>
</li>
<li>Analysis - survival analysis<ul>
<li>Associations between inferred proportions of immune cell types and survival were tested using <strong>Cox regression</strong></li>
</ul>
</li>
</ul>
<h2 id="003-Profiles-of-immune-cell-infiltration-and-immune-related-genes-in-the-tumor-microenvironment-of-colorectal-cancer"><a href="#003-Profiles-of-immune-cell-infiltration-and-immune-related-genes-in-the-tumor-microenvironment-of-colorectal-cancer" class="headerlink" title="003 Profiles of immune cell infiltration and immune-related genes in the tumor microenvironment of colorectal cancer"></a>003 Profiles of immune cell infiltration and immune-related genes in the tumor microenvironment of colorectal cancer</h2><p>本地：<a href="BItest2/BItest2_3.pdf">BItest2_3</a><br>期刊：Biomed Pharmacother, 2, 7.419, 2019<br>结直肠癌肿瘤微环境中免疫细胞浸润和免疫相关基因的概况</p>
<h3 id="DATA-AVAIABLE"><a href="#DATA-AVAIABLE" class="headerlink" title="DATA AVAIABLE"></a>DATA AVAIABLE</h3><ul>
<li>GSE21510</li>
</ul>
<h3 id="Process"><a href="#Process" class="headerlink" title="Process"></a>Process</h3><p><strong>CIBERSORT,</strong> a deconvolution algorithm, was used to analyze the infiltration of 22 immune cell types in the tumor microenvironment and immune-related gene expression in 404 colorectal cancer (CRC). and 40 adjacent non-tumorous tissues</p>
<ul>
<li>edge R包进行差异表达的筛选<ul>
<li>with parameters of logFoldchange (FC) &gt; 2 or &lt; -2 and P value &lt; 0.05.</li>
</ul>
</li>
<li>CIBERSOFT</li>
<li>survival analysis<ul>
<li>CIBERSORT p值&lt; 0.05的病例纳入后续分析</li>
<li>用Wilcox试验评价TCGA组和GEO组22种免疫细胞的浸润差异</li>
<li>用corrplot R包检测不同免疫细胞间的相关性</li>
<li>Kaplan-Meier曲线分析采用生存R包</li>
<li>采用Cox回归进行单变量和多变量分析</li>
<li>ROC分析采用时间ROC R包计算生存率。</li>
<li>采用survcomp R软件包评定一致性指数。</li>
<li>IL23A和CD68的表达数据采用双尾配对t检验。</li>
<li>采用Wilcoxon标记秩检验对GEO队列中CRC与相邻非肿瘤组织之间的免疫相关基因进行验证。</li>
</ul>
</li>
</ul>
<h2 id="conclusion"><a href="#conclusion" class="headerlink" title="conclusion"></a>conclusion</h2><p>However, past work has not accounted for the <strong>diversity of functionally distinct cell types</strong> that make up the immune response. The aim of this study was to determine whether differences in the cellular composition of the immune infiltrate in xxx cancer</p>
<ul>
<li>data pre-processing<ul>
<li>包括将转录组数据转化为微阵列？</li>
<li>冗杂数据的剔除</li>
<li>差异分析的筛选</li>
</ul>
</li>
<li>process<ul>
<li>CIBERSOFT</li>
<li>survival analysis <ul>
<li>cox</li>
<li>Hierarchical clustering</li>
<li>Kaplan- Meier curves</li>
<li>log- rank test， t-test</li>
</ul>
</li>
</ul>
</li>
</ul>
<h1 id="Test2-for-differential-analysis"><a href="#Test2-for-differential-analysis" class="headerlink" title="Test2 - for differential analysis"></a>Test2 - for differential analysis</h1><h2 id="001-Identification-of-potential-key-genes-for-HER-2-positive-breast-cancer-based-on-bioinformatics-analysis"><a href="#001-Identification-of-potential-key-genes-for-HER-2-positive-breast-cancer-based-on-bioinformatics-analysis" class="headerlink" title="001 Identification of potential key genes for HER-2 positive breast cancer based on bioinformatics analysis"></a>001 Identification of potential key genes for HER-2 positive breast cancer based on bioinformatics analysis</h2><p>medicine(Baltimore): 4, 1.817, 2020</p>
<p>基于生物信息学分析的HER-2阳性乳腺癌潜在关键基因的鉴定</p>
<ul>
<li>HER-2阳性乳腺癌是一种临床转归较差的乳腺癌亚型。本研究的目的是鉴定HER-2阳性乳腺癌的差异表达基因(DEGs)，并阐明它们之间潜在的相互作用。</li>
<li>数据集<ul>
<li>GSE29431、GSE45827和GSE65194</li>
</ul>
</li>
<li>流程<ul>
<li>应用<strong>GEO2R工具</strong>获取HER-2阳性乳腺癌与正常乳腺组织之间的差异基因。</li>
<li><strong>基因本体(GO)注释分析</strong>和<strong>京都基因与基因组百科全书(KEGG)路径富集分析</strong>采用注释、可视化和集成发现数据库 <strong>(David)在线工具</strong>进行。</li>
<li>利用<strong>Cytoscape软件</strong>进行蛋白-蛋白相互作用(Protein-protein interaction, <strong>PPI</strong>)网络、<strong>hub基因鉴定</strong>和<strong>模块分析</strong>。</li>
<li>在线<strong>Kaplan-Meier plotter生存分析工具</strong>也用于研究枢纽基因在HER-2阳性乳腺癌患者中的预后价值。</li>
</ul>
</li>
<li>结果<ul>
<li>共鉴定出54个上调的DEGs和269个下调的DEGs。其中筛选出10个在PPI网络中具有连通程度的枢纽基因，包括CCNB1、RAC1、TOP2A、KIF20A、RRM2、ASPM、NUSAP1、BIRC5、BUB1B和CEP55。Kaplan-Meier绘图生存分析显示，HER-2阳性乳腺癌患者RAC1和RRM2过表达与不良预后相关。</li>
</ul>
</li>
<li>结论<ul>
<li>本研究发现了一些可能影响HER-2阳性乳腺癌发生和进展的潜在靶基因和通路。这些发现可以为这种疾病的新型诊断和治疗生物标志物的检测提供新的见解。</li>
</ul>
</li>
</ul>
<h2 id="002-Identification-of-potential-core-genes-in-triple-negative-breast-cancer-using-bioinformatics-analysis"><a href="#002-Identification-of-potential-core-genes-in-triple-negative-breast-cancer-using-bioinformatics-analysis" class="headerlink" title="002 Identification of potential core genes in triple negative breast cancer using bioinformatics analysis"></a>002 Identification of potential core genes in triple negative breast cancer using bioinformatics analysis</h2><p>Onco Targets Ther: 4, 4.345, 2018<br>三阴性乳腺癌潜在核心基因的生物信息学分析</p>
<ul>
<li>三阴性乳腺癌(TNBC)是一种临床预后差、治疗选择有限的乳腺癌亚型。由于缺乏分子靶点，化疗是TNBC患者的主要辅助治疗方法。</li>
<li>data<ul>
<li>三个微阵列数据集(GSE38959、GSE45827和GSE65194)</li>
</ul>
</li>
<li>process<ul>
<li>利用<strong>GEO2R</strong>工具筛选出TNBC与正常组织之间的差异表达基因(DEGs)。</li>
<li>利用该数据库进行了基因本体功能和京都基因与基因组百科的路径富集分析确定DEGs的路径和功能注释。</li>
<li>基于互作基因检索工具(Search Tool for Retrieval of interaction Genes)数据库，利用<strong>Cytoscape软件</strong>进行可视化分析，分析这些DEGs的蛋白-蛋白相互作用。</li>
<li>使用在线<strong>Kaplan-Meier plotter生存分析工具</strong>评估乳腺癌患者枢纽基因表达的预后价值。</li>
</ul>
</li>
<li>result:<ul>
<li>共鉴定到278个表达上调的DEGs和173个表达下调的DEGs。从中筛选出10个连接度较高的枢纽基因。这些枢纽基因的过表达与乳腺癌预后不良相关，特别是CCNB1的过表达提示TNBC预后不良。</li>
</ul>
</li>
<li>conclusion<ul>
<li>与正常乳腺组织相比，CCNB1在TNBC中过表达，CCNB1过表达是TNBC患者预后的不利因素。CCNB1在TNBC治疗中的价值有待进一步研究。</li>
</ul>
</li>
</ul>
<h2 id="003-Identification-of-core-genes-and-potential-molecular-mechanisms-in-breast-cancer-using-bioinformatics-analysis"><a href="#003-Identification-of-core-genes-and-potential-molecular-mechanisms-in-breast-cancer-using-bioinformatics-analysis" class="headerlink" title="003 Identification of core genes and potential molecular mechanisms in breast cancer using bioinformatics analysis"></a>003 Identification of core genes and potential molecular mechanisms in breast cancer using bioinformatics analysis</h2><p>Pathol Res Pract: 4, 3.309, 2019<br>利用生物信息学分析鉴定乳腺癌的核心基因和潜在分子机制</p>
<ul>
<li>乳腺癌是世界范围内最常见的女性癌症。本研究旨在阐明乳腺癌潜在的关键候选基因和途径。</li>
<li>data<ul>
<li>GSE65212</li>
</ul>
</li>
<li>prcess<ul>
<li>差异表达基因(DEGs)由<strong>R Bioconductor包</strong>获得。</li>
<li>利用<strong>DAVID数据库</strong>对DEGs进行基因本体论(GO)和京都基因与基因组百科(KEGG)途径富集分析。</li>
<li>利用<strong>STRING</strong>建立蛋白-蛋白相互作用(protein-protein interaction, PPI)网络，利用<strong>Cytoscape软件</strong>进行可视化分析。</li>
<li>用<strong>插件式分子复合体检测(MCODE)</strong> 对PPI网络进行模块分析。</li>
<li>利用<strong>Kaplan-Meier绘图仪</strong>在线数据库和<strong>qPCR</strong> (quantitative real-time PCR)对乳腺癌组织样本进行基因鉴定。</li>
</ul>
</li>
</ul>
<h2 id="004-Identification-of-breast-cancer-hub-genes-and-analysis-of-prognostic-values-using-integrated-bioinformatics-analysis"><a href="#004-Identification-of-breast-cancer-hub-genes-and-analysis-of-prognostic-values-using-integrated-bioinformatics-analysis" class="headerlink" title="004 Identification of breast cancer hub genes and analysis of prognostic values using integrated bioinformatics analysis"></a>004 Identification of breast cancer hub genes and analysis of prognostic values using integrated bioinformatics analysis</h2><p>Cancer Biomark: 3, 3.828, 2017<br>利用综合生物信息学分析乳腺癌中枢基因的鉴定及预后价值分析</p>
<ul>
<li>乳腺癌(BC)是美国女性癌症死亡的第二大常见原因。由于BC的分子机制尚未完全被发现，因此对该疾病中枢基因和通路的识别对于揭示乳腺癌发生发展的分子机制具有重要意义。本研究旨在寻找BC治疗中潜在的生物标志物和hub基因的生存分析。</li>
<li>DATA<ul>
<li>GSE10797</li>
</ul>
</li>
<li>process<ul>
<li>利用基因表达综合(GEO)数据库中的微阵列数据筛选乳腺癌细胞与正常细胞之间的**差异表达基因(DEGs)**。</li>
<li>利用<strong>DAVID数据库</strong>对DEGs进行<strong>基因本体(GO)和KEGG途径富集分析</strong>，</li>
<li>利用<strong>Cytoscape软件</strong>构建蛋白质-蛋白质相互作用**(PPI)网络**，</li>
<li>利用<strong>MCODE</strong>进行模块分析。</li>
<li>然后，利用<strong>Kaplan-Meier绘图仪</strong>在线工具对轮毂基因进行<strong>总体存活(OS)分析。</strong></li>
<li>最后，利用<strong>连通性图(Connectivity Map, cMap)数据库</strong>对潜在的分子制剂进行了鉴定。</li>
</ul>
</li>
<li>coclusion<ul>
<li>由于程度高，选取了15个基因作为枢纽基因，其中RPS9、RPL11、RPS14、RPL10A 4个基因的低表达与BC患者OS较差相关。此外，小分子药物伊美汀可能是一种潜在的治疗BC的药物。</li>
</ul>
</li>
</ul>
<h2 id="005-KEGG-expressed-genes-and-pathways-in-triple-negative-breast-cancer-Protocol-for-a-systematic-review-and-data-mining"><a href="#005-KEGG-expressed-genes-and-pathways-in-triple-negative-breast-cancer-Protocol-for-a-systematic-review-and-data-mining" class="headerlink" title="005 KEGG-expressed genes and pathways in triple negative breast cancer: Protocol for a systematic review and data mining"></a>005 KEGG-expressed genes and pathways in triple negative breast cancer: Protocol for a systematic review and data mining</h2><p>Medicine (Baltimore)<br>三阴性乳腺癌中kegg表达基因和途径:系统综述和数据挖掘的方案</p>
<ul>
<li>三阴性乳腺癌(TNBC)的发病率处于较高水平，我们的研究旨在识别TNBC中的差异表达基因(DEGs)，并探索TNBC的关键途径和关键基因。</li>
<li>data<ul>
<li>GSE86945、GSE86946和GSE102088</li>
</ul>
</li>
<li>process<ul>
<li>基因表达谱(DEGs)采用R软件进行识别，基因本体(GO)分析和京都基因与基因组百科(KEGG)路径富集分析采用<strong>DAVID</strong> (Database for Annotation、Visualization和Integrated Discovery)工具进行。</li>
<li>利用STRING数据库构建蛋白质相互作用(protein-protein interaction, PPI)网络，</li>
<li>利用Cytoscape软件进行可视化分析。</li>
<li>最后，利用Kaplan-Meier绘图仪在线工具对乳腺癌患者中hub DEGs的生存价值进行测定。</li>
</ul>
</li>
</ul>
<h2 id="Conclusion-1"><a href="#Conclusion-1" class="headerlink" title="Conclusion"></a>Conclusion</h2><ul>
<li>data acquisition</li>
<li>DEGs</li>
<li>GO分析和富集分析<ul>
<li>DAVID</li>
</ul>
</li>
<li>PIP网络<ul>
<li>STRING</li>
</ul>
</li>
<li>数据可视化<ul>
<li>Cytoscape</li>
</ul>
</li>
<li>预后价值测定<ul>
<li>Kaplan-Meier绘图仪</li>
</ul>
</li>
</ul>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="http://example.com">pigudog</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://example.com/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/">http://example.com/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest2/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/BItest/">BItest</a></div><div class="post_share"><div class="social-share" data-image="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><div class="post-reward"><div class="reward-button"><i class="fas fa-qrcode"></i> Donate</div><div class="reward-main"><ul class="reward-all"><li class="reward-item"><a href="/img/wechat.jpg" target="_blank"><img class="post-qr-code-img" src="/img/wechat.jpg" alt="wechat"/></a><div class="post-qr-code-desc">wechat</div></li></ul></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest1/"><img class="prev-cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">BItest1 - 学会解析文章架构</div></div></a></div><div class="next-post pull-right"><a href="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA6/"><img class="next-cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">Introduction to WES</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest1/" title="BItest1 - 学会解析文章架构"><img class="cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2022-11-17</div><div class="title">BItest1 - 学会解析文章架构</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/img/avatar.gif" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">pigudog</div><div class="author-info__description">冲冲冲！</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">70</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">12</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">16</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/pigudog"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/pigudog" target="_blank" title="Github"><i class="fab fa-github"></i></a><a class="social-icon" href="/2205638116@qq.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#Quetion"><span class="toc-number">1.</span> <span class="toc-text">Quetion</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Test1-for-CIBERSOFT"><span class="toc-number">2.</span> <span class="toc-text">Test1 - for CIBERSOFT</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%82%BF%E7%98%A4%E5%85%8D%E7%96%AB%E5%BE%AE%E7%8E%AF%E5%A2%83%E5%92%8C%E8%82%BF%E7%98%A4%E5%85%8D%E7%96%AB%E6%B5%B8%E6%B6%A6"><span class="toc-number">2.1.</span> <span class="toc-text">肿瘤免疫微环境和肿瘤免疫浸润</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#CIBRSOFT"><span class="toc-number">2.2.</span> <span class="toc-text">CIBRSOFT</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BB%80%E4%B9%88%E6%98%AFCIBERSOFT"><span class="toc-number">2.2.1.</span> <span class="toc-text">什么是CIBERSOFT</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#HOW-TO-USE-CIBERSOFT"><span class="toc-number">2.2.2.</span> <span class="toc-text">HOW TO USE CIBERSOFT</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Cibersoft-Data-visualization"><span class="toc-number">2.2.3.</span> <span class="toc-text">Cibersoft Data visualization</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#001-Profiles-of-immune-infiltration-in-colorectal-cancer-and-their-clinical-significant-A-gene-expression%E2%80%90based-study"><span class="toc-number">2.3.</span> <span class="toc-text">001 Profiles of immune infiltration in colorectal cancer and their clinical significant: A gene expression‐based study</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Available-of-data-and-materials"><span class="toc-number">2.3.1.</span> <span class="toc-text">Available of data and materials</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Conclusion"><span class="toc-number">2.3.2.</span> <span class="toc-text">Conclusion</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Noun-explanation"><span class="toc-number">2.3.3.</span> <span class="toc-text">Noun explanation</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Analysis-of-Process"><span class="toc-number">2.3.4.</span> <span class="toc-text">Analysis of Process</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#002-Patterns-of-Immune-Infiltration-in-Breast-Cancer-and-Their-Clinical-Implications-A-Gene-Expression-Based-Retrospective-Study"><span class="toc-number">2.4.</span> <span class="toc-text">002 Patterns of Immune Infiltration in Breast Cancer and Their Clinical Implications: A Gene-Expression-Based Retrospective Study</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Conclustion"><span class="toc-number">2.4.1.</span> <span class="toc-text">Conclustion</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#available-of-data"><span class="toc-number">2.4.2.</span> <span class="toc-text">available of data</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Analysis-of-process"><span class="toc-number">2.4.3.</span> <span class="toc-text">Analysis of process</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#003-Profiles-of-immune-cell-infiltration-and-immune-related-genes-in-the-tumor-microenvironment-of-colorectal-cancer"><span class="toc-number">2.5.</span> <span class="toc-text">003 Profiles of immune cell infiltration and immune-related genes in the tumor microenvironment of colorectal cancer</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#DATA-AVAIABLE"><span class="toc-number">2.5.1.</span> <span class="toc-text">DATA AVAIABLE</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Process"><span class="toc-number">2.5.2.</span> <span class="toc-text">Process</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#conclusion"><span class="toc-number">2.6.</span> <span class="toc-text">conclusion</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#Test2-for-differential-analysis"><span class="toc-number">3.</span> <span class="toc-text">Test2 - for differential analysis</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#001-Identification-of-potential-key-genes-for-HER-2-positive-breast-cancer-based-on-bioinformatics-analysis"><span class="toc-number">3.1.</span> <span class="toc-text">001 Identification of potential key genes for HER-2 positive breast cancer based on bioinformatics analysis</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#002-Identification-of-potential-core-genes-in-triple-negative-breast-cancer-using-bioinformatics-analysis"><span class="toc-number">3.2.</span> <span class="toc-text">002 Identification of potential core genes in triple negative breast cancer using bioinformatics analysis</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#003-Identification-of-core-genes-and-potential-molecular-mechanisms-in-breast-cancer-using-bioinformatics-analysis"><span class="toc-number">3.3.</span> <span class="toc-text">003 Identification of core genes and potential molecular mechanisms in breast cancer using bioinformatics analysis</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#004-Identification-of-breast-cancer-hub-genes-and-analysis-of-prognostic-values-using-integrated-bioinformatics-analysis"><span class="toc-number">3.4.</span> <span class="toc-text">004 Identification of breast cancer hub genes and analysis of prognostic values using integrated bioinformatics analysis</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#005-KEGG-expressed-genes-and-pathways-in-triple-negative-breast-cancer-Protocol-for-a-systematic-review-and-data-mining"><span class="toc-number">3.5.</span> <span class="toc-text">005 KEGG-expressed genes and pathways in triple negative breast cancer: Protocol for a systematic review and data mining</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Conclusion-1"><span class="toc-number">3.6.</span> <span class="toc-text">Conclusion</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2023/09/18/pigudog/pigudog1-palantir/" title="pigudog1-palantir"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="pigudog1-palantir"/></a><div class="content"><a class="title" href="/2023/09/18/pigudog/pigudog1-palantir/" title="pigudog1-palantir">pigudog1-palantir</a><time datetime="2023-09-17T16:00:00.000Z" title="Created 2023-09-18 00:00:00">2023-09-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/09/18/pigudog/pigudog2-palantir/" title="pigudog2-palantir公众号版本"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="pigudog2-palantir公众号版本"/></a><div class="content"><a class="title" href="/2023/09/18/pigudog/pigudog2-palantir/" title="pigudog2-palantir公众号版本">pigudog2-palantir公众号版本</a><time datetime="2023-09-17T16:00:00.000Z" title="Created 2023-09-18 00:00:00">2023-09-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/16/statistic/statistic7%20-%20Analysis%20of%20variance%20and%20regression/" title="statistic7 - Analysis of variance and regression"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic7 - Analysis of variance and regression"/></a><div class="content"><a class="title" href="/2023/02/16/statistic/statistic7%20-%20Analysis%20of%20variance%20and%20regression/" title="statistic7 - Analysis of variance and regression">statistic7 - Analysis of variance and regression</a><time datetime="2023-02-15T16:00:00.000Z" title="Created 2023-02-16 00:00:00">2023-02-16</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/15/statistic/statistic6%20-%20Hypothesis%20testing/" title="statistic6 - Hypothesis testing"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic6 - Hypothesis testing"/></a><div class="content"><a class="title" href="/2023/02/15/statistic/statistic6%20-%20Hypothesis%20testing/" title="statistic6 - Hypothesis testing">statistic6 - Hypothesis testing</a><time datetime="2023-02-14T16:00:00.000Z" title="Created 2023-02-15 00:00:00">2023-02-15</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/13/statistic/statistic4%20-%20Sampling/" title="statistic4 - Sampling"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic4 - Sampling"/></a><div class="content"><a class="title" href="/2023/02/13/statistic/statistic4%20-%20Sampling/" title="statistic4 - Sampling">statistic4 - Sampling</a><time datetime="2023-02-12T16:00:00.000Z" title="Created 2023-02-13 00:00:00">2023-02-13</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2022 - 2023 By pigudog</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div><div class="footer_custom_text">Hi, welcome to my <a target="_blank" rel="noopener" href="https://pigudog.com">blog</a>!</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="Switch Between Traditional Chinese And Simplified Chinese">简</button><button id="darkmode" type="button" title="Switch Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><div class="js-pjax"></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>