<!DOCTYPE html><html lang="en" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>pigudog1-palantir | PiguDog的小屋</title><meta name="keywords" content="-medical -computer"><meta name="author" content="pigudog,2205638116@qq.com"><meta name="copyright" content="pigudog"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="IntroductionPalantir is an algorithm to align cells along differentiation trajectories. Palantir models differentiation as a stochastic process where stem cells differentiate to terminally differentia">
<meta property="og:type" content="article">
<meta property="og:title" content="pigudog1-palantir">
<meta property="og:url" content="http://example.com/2023/09/18/pigudog/pigudog1-palantir/index.html">
<meta property="og:site_name" content="PiguDog的小屋">
<meta property="og:description" content="IntroductionPalantir is an algorithm to align cells along differentiation trajectories. Palantir models differentiation as a stochastic process where stem cells differentiate to terminally differentia">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png">
<meta property="article:published_time" content="2023-09-17T16:00:00.000Z">
<meta property="article:modified_time" content="2023-09-18T11:42:15.573Z">
<meta property="article:author" content="pigudog">
<meta property="article:tag" content="-medical -computer">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="http://example.com/2023/09/18/pigudog/pigudog1-palantir/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: {"defaultEncoding":1,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: 'days',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: {"limitCount":50,"languages":{"author":"Author: pigudog","link":"Link: ","source":"Source: PiguDog的小屋","info":"Copyright is owned by the author. For commercial reprints, please contact the author for authorization. For non-commercial reprints, please indicate the source."}},
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'pigudog1-palantir',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2023-09-18 19:42:15'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.2.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/img/avatar.gif" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">69</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">12</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">16</div></a></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> Movie</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">PiguDog的小屋</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> Movie</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">pigudog1-palantir<a class="post-edit-link" href="https://github.com/pigudog/_posts/pigudog/pigudog1-palantir.md" title="Edited on" target="_blank"><i class="fas fa-pencil-alt"></i></a></h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2023-09-17T16:00:00.000Z" title="Created 2023-09-18 00:00:00">2023-09-18</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2023-09-18T11:42:15.573Z" title="Updated 2023-09-18 19:42:15">2023-09-18</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="pigudog1-palantir"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="Introduction"><a href="#Introduction" class="headerlink" title="Introduction"></a>Introduction</h1><p><img src="/2023/09/18/pigudog/pigudog1-palantir/Pasted%20image%2020230918102756.png"><br>Palantir is an algorithm to align cells along differentiation trajectories. Palantir models differentiation as a stochastic process where stem cells differentiate to terminally differentiated cells by a series of steps through a low dimensional phenotypic manifold. Palantir effectively captures the continuity in cell states and the stochasticity in cell fate determination.</p>
<p>Palantir 是通过随机过程，低维流型设计的拟时序分析算法，可以有效地捕捉细胞的性状和细胞命运决定的随机性，并基于以下两个基本的假设</p>
<ul>
<li>Firstly, as in all pseudo-time inference algorithms, we assume unidirectional progression from a less- to a more-differentiated state. We posit that it is a reasonable first order approximation for healthy differentiation. 也就是说这个模型假设了细胞分化用的是一阶近似，但是要注意到对于癌症中可能会出现的包括mutation等特殊情况，可能就不适用。但是如果我们对月经周期，妊娠周期等以时间变化的健康模型可能会有不错的效果</li>
<li>Second, we assume that for any node, the probability of traversing to neighbor is independent of its history, that is, the cell’s development history is likely to be encoded in its epigenetic profile and will probably impact cell fate choice.  However, nodes are abstract cell states representing multiple histories and potential trajectories rather than the path of an individual cell. 简单地说，通过宏观的角度看，我们认为每个空间位置的细胞的趋化只与自己相关，这也就引出了markov模型</li>
</ul>
<h2 id="Markov-chain"><a href="#Markov-chain" class="headerlink" title="Markov chain"></a>Markov chain</h2><p>我们先来了解一下markov chain：</p>
<ul>
<li>马尔可夫链是一系列具有有限或可数结果的随机事件，其中每个事件发生的概率仅取决于前一个事件达到的状态。它的特点是，笼统地说，在固定的现在下，未来独立于过去。</li>
<li>马尔科夫链为状态空间中经过从一个状态到另一个状态的转换的随机过程，该过程要求具备“无记忆性 ”，即下一状态的概率分布只能由当前状态决定，在时间序列中它前面的事件均与之无关。这种特定类型的“无记忆性 ”称作马尔可夫性质 – 也就是我们palantir的第二假设<ul>
<li>即$P(x_t|x_0,x_1,…x_{t-1})&#x3D;P(x_t|x_{t-1})$</li>
</ul>
</li>
<li>离散马尔科夫链<ul>
<li>这个其实在我们的rna文库比对上就有用到，主要是就是通过一个转移概率矩阵，最终得到收敛的结果</li>
<li>设马尔科夫链$X&#x3D;{x_0,x_1,…,x_i}$，其状态空间为S，转移矩阵为$P&#x3D;{p_{ij}}$</li>
<li>我们先假设我们的马尔科夫链存在唯一的分布<ul>
<li>也就是$\begin{cases}x_i&#x3D;\sum{p_{ij}x_j}\x_i&gt;0\ \sum{x_i}&#x3D;1\end{cases}$存在一组解</li>
</ul>
</li>
</ul>
</li>
<li>而我们的palantir就是基于每个spot的细胞，它的分化的方向就是一个转移，通过得到它的转移概率矩阵，计算细胞的分化过程</li>
</ul>
<h2 id="查看源码"><a href="#查看源码" class="headerlink" title="查看源码"></a>查看源码</h2><p>我们看一下palantir使用markov chain对细胞的分化进行了研究的代码</p>
<ul>
<li>它接受了多个输入参数，包括包括<code>wp_data</code>、<code>knn</code>、<code>pseudotime</code>和<code>n_jobs</code></li>
<li>它使用scikit-learn库中的<code>NearestNeighbors</code>类构建了一个k最近邻（kNN）图。kNN图表示了<code>wp_data</code>中数据点之间的连接关系</li>
<li>计算了自适应k值的标准差，允许”back” edge 在kNN图中存在<ul>
<li>自适应标准差(adaptive standard deviation)是一种标准差，在这里主要是指通过<code>n_neighbors</code>参数对于标准差做了调整，因为如果你的最近邻的个数多，其实意味着你的相似性就更高，所以也就能构建我们的亲和力矩阵</li>
</ul>
</li>
<li>kNN图中删除在伪时间中向后移动的边，但保留在计算的标准差内的边，得到新的KNN邻接矩阵</li>
<li>构建亲和力矩阵这一步很关键<ul>
<li><code>x, y, z = find(kNN)</code>: 这行代码从已构建的kNN图中提取了三个重要的参数，分别是x（源节点的索引）、y（目标节点的索引）和z（它们之间的距离或相似性）</li>
<li>亲和力矩阵中的每个元素表示两个路径点之间的相似性。计算过程是基于高斯核函数，它使用了kNN图中的距离信息（z）以及自适应标准差（adaptive_std）来度量相似性。<figure class="highlight py"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">def</span> <span class="title function_">_construct_markov_chain</span>(<span class="params">wp_data, knn, pseudotime, n_jobs</span>):</span><br><span class="line">    <span class="comment"># Markov chain construction</span></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&quot;Markov chain construction...&quot;</span>)</span><br><span class="line">    waypoints = wp_data.index</span><br><span class="line"></span><br><span class="line">    <span class="comment"># kNN graph 构建KNN图</span></span><br><span class="line">    n_neighbors = knn</span><br><span class="line">    nbrs = NearestNeighbors(</span><br><span class="line">        n_neighbors=n_neighbors, metric=<span class="string">&quot;euclidean&quot;</span>, n_jobs=n_jobs</span><br><span class="line">    ).fit(wp_data)</span><br><span class="line">    kNN = nbrs.kneighbors_graph(wp_data, mode=<span class="string">&quot;distance&quot;</span>)</span><br><span class="line">    dist, ind = nbrs.kneighbors(wp_data)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Standard deviation allowing for &quot;back&quot; edges</span></span><br><span class="line">    adpative_k = np.<span class="built_in">min</span>([<span class="built_in">int</span>(np.floor(n_neighbors / <span class="number">3</span>)) - <span class="number">1</span>, <span class="number">30</span>])</span><br><span class="line">    adaptive_std = np.ravel(dist[:, adpative_k])</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Directed graph construction它确定了所有邻居在伪时间中的位置，针对输入数据中的每个路径点</span></span><br><span class="line">    <span class="comment"># pseudotime position of all the neighbors</span></span><br><span class="line">    traj_nbrs = pd.DataFrame(</span><br><span class="line">        pseudotime[np.ravel(waypoints.values[ind])].values.reshape(</span><br><span class="line">            [<span class="built_in">len</span>(waypoints), n_neighbors]</span><br><span class="line">        ),</span><br><span class="line">        index=waypoints,</span><br><span class="line">    )</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Remove edges that move backwards in pseudotime except for edges that are within它从kNN图中删除在伪时间中向后移动的边，但保留在计算的标准差内的边</span></span><br><span class="line">    <span class="comment"># the computed standard deviation</span></span><br><span class="line">    rem_edges = traj_nbrs.apply(</span><br><span class="line">        <span class="keyword">lambda</span> x: x &lt; pseudotime[traj_nbrs.index] - adaptive_std</span><br><span class="line">    )</span><br><span class="line">    rem_edges = rem_edges.stack()[rem_edges.stack()]</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Determine the indices and update adjacency matrix它根据删除的边更新了kNN图的邻接矩阵</span></span><br><span class="line">    cell_mapping = pd.Series(<span class="built_in">range</span>(<span class="built_in">len</span>(waypoints)), index=waypoints)</span><br><span class="line">    x = <span class="built_in">list</span>(cell_mapping[rem_edges.index.get_level_values(<span class="number">0</span>)])</span><br><span class="line">    y = <span class="built_in">list</span>(rem_edges.index.get_level_values(<span class="number">1</span>))</span><br><span class="line">    <span class="comment"># Update adjacecy matrix</span></span><br><span class="line">    kNN[x, ind[x, y]] = <span class="number">0</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># Affinity matrix and markov chain它基于更新后的kNN图计算了一个亲和力矩阵，用于量化路径点之间的相似性</span></span><br><span class="line">    x, y, z = find(kNN)</span><br><span class="line">    aff = np.exp(</span><br><span class="line">        -(z**<span class="number">2</span>) / (adaptive_std[x] ** <span class="number">2</span>) * <span class="number">0.5</span></span><br><span class="line">        - (z**<span class="number">2</span>) / (adaptive_std[y] ** <span class="number">2</span>) * <span class="number">0.5</span></span><br><span class="line">    )</span><br><span class="line">    W = csr_matrix((aff, (x, y)), [<span class="built_in">len</span>(waypoints), <span class="built_in">len</span>(waypoints)])</span><br><span class="line"></span><br><span class="line">    <span class="comment"># Transition matrix根据亲和力矩阵和路径点的度矩阵构建了一个转移矩阵</span></span><br><span class="line">    D = np.ravel(W.<span class="built_in">sum</span>(axis=<span class="number">1</span>))</span><br><span class="line">    x, y, z = find(W)</span><br><span class="line">    T = csr_matrix((z / D[x], (x, y)), [<span class="built_in">len</span>(waypoints), <span class="built_in">len</span>(waypoints)])</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> T <span class="comment"># 最后，它返回转移矩阵`T`，表示马尔可夫链</span></span><br></pre></td></tr></table></figure></li>
</ul>
</li>
</ul>
<h1 id="preparatory-work"><a href="#preparatory-work" class="headerlink" title="preparatory work"></a>preparatory work</h1><ul>
<li>Paper： <a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41587-019-0068-4">https://www.nature.com/articles/s41587-019-0068-4</a></li>
<li>github： <a target="_blank" rel="noopener" href="https://github.com/dpeerlab/Palantir">https://github.com/dpeerlab/Palantir</a></li>
<li>tutorial： <a target="_blank" rel="noopener" href="https://nbviewer.org/github/dpeerlab/Palantir/blob/master/notebooks/Palantir_sample_notebook.ipynb">https://nbviewer.org/github/dpeerlab/Palantir/blob/master/notebooks/Palantir_sample_notebook.ipynb</a></li>
<li>datasets： <a target="_blank" rel="noopener" href="https://github.com/dpeerlab/Palantir/tree/master/data">https://github.com/dpeerlab/Palantir/tree/master/data</a></li>
</ul>
<p>因为palantir包使用过程中可能会出现语言文字的不存在的报错，我们可以直接寻找matplotlib中在我们的系统中自带的font</p>
<ul>
<li><img src="/2023/09/18/pigudog/pigudog1-palantir/Pasted%20image%2020230918180729.png"><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> font_manager</span><br><span class="line"><span class="keyword">for</span> font <span class="keyword">in</span> font_manager.fontManager.ttflist[:<span class="number">5</span>]:</span><br><span class="line">        <span class="comment"># 查看字体名以及对应的字体文件名</span></span><br><span class="line">    <span class="comment"># print(font.name, &#x27;-&#x27;, font.fname)</span></span><br><span class="line">     <span class="built_in">print</span>(font.name)</span><br></pre></td></tr></table></figure>

  STIXGeneral<br>  DejaVu Sans Mono<br>  cmss10<br>  DejaVu Sans<br>  DejaVu Sans Mono</li>
</ul>
<p>为了更好地理解palantir的运行方式以及对一些warning的问题进行调试，我们将palantir包下载并命名为<code>palantir_modify</code></p>
<ul>
<li><img src="/2023/09/18/pigudog/pigudog1-palantir/Pasted%20image%2020230918181113.png"><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> palantir_modify <span class="keyword">as</span> pa</span><br></pre></td></tr></table></figure></li>
</ul>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line"><span class="comment"># Plotting</span></span><br><span class="line"><span class="keyword">import</span> matplotlib</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"></span><br><span class="line"><span class="comment"># warnings</span></span><br><span class="line"><span class="keyword">import</span> warnings</span><br><span class="line"><span class="keyword">from</span> numba.core.errors <span class="keyword">import</span> NumbaDeprecationWarning</span><br><span class="line"></span><br><span class="line">warnings.filterwarnings(action=<span class="string">&quot;ignore&quot;</span>, category=NumbaDeprecationWarning)</span><br><span class="line">warnings.filterwarnings(</span><br><span class="line">    action=<span class="string">&quot;ignore&quot;</span>, module=<span class="string">&quot;scanpy&quot;</span>, message=<span class="string">&quot;No data for colormapping&quot;</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">plt.rcParams[<span class="string">&#x27;font.family&#x27;</span>] = <span class="string">&quot;DejaVu Sans&quot;</span></span><br><span class="line"><span class="comment"># Inline plotting</span></span><br><span class="line">%matplotlib inline</span><br></pre></td></tr></table></figure>

<p>为了更好的复现我们的结果，我们对随机参数进行设定，在palantir包中主要是使用numpy包进行set seed</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> random</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line">random_seed = <span class="number">1314</span></span><br><span class="line">random.seed(random_seed)</span><br><span class="line">np.random.seed(random_seed)</span><br></pre></td></tr></table></figure>

<h1 id="1-load-and-preprocess"><a href="#1-load-and-preprocess" class="headerlink" title="1. load and preprocess"></a>1. load and preprocess</h1><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Load sample data</span></span><br><span class="line">data_dir = os.path.expanduser(<span class="string">&quot;./data/&quot;</span>)</span><br><span class="line">download_url = <span class="string">&quot;https://dp-lab-data-public.s3.amazonaws.com/palantir/marrow_sample_scseq_counts.h5ad&quot;</span></span><br><span class="line">file_path = os.path.join(data_dir, <span class="string">&quot;marrow_sample_scseq_counts.h5ad&quot;</span>)</span><br><span class="line">ad = sc.read(file_path, backup_url=download_url)</span><br><span class="line">ad</span><br></pre></td></tr></table></figure>
<pre><code>AnnData object with n_obs × n_vars = 4142 × 16106
</code></pre>
<p>这是一个count矩阵，保留整数的稀疏矩阵</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ad.X.<span class="built_in">max</span>()</span><br></pre></td></tr></table></figure>
<pre><code>632.0
</code></pre>
<h1 id="2-preprocess"><a href="#2-preprocess" class="headerlink" title="2. preprocess"></a>2. preprocess</h1><p>palantir包中的<code>prerpocess</code>模块中主要包括以下的三个函数</p>
<ul>
<li>&#96;filter_counts_data(data, cell_min_molecules&#x3D;1000, genes_min_cells&#x3D;10)&#96;&#96;</li>
<li>&#96;normalize_counts(data)&#96;&#96;</li>
<li><code>log_transform(data, pseudo_count=0.1)`` 我们使用</code>normalize_total<code>进行标准化，注意</code>normalize_per_cell&#96;是老版本并且依旧可用<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># sc.pp.normalize_per_cell(ad)</span></span><br><span class="line"><span class="comment"># after normalization, each observation (cell) has a total count equal to the median of total counts for observations (cells) before normalization.</span></span><br><span class="line">sc.pp.normalize_total(ad)</span><br><span class="line">ad</span><br></pre></td></tr></table></figure>
  AnnData object with n_obs × n_vars &#x3D; 4142 × 16106</li>
</ul>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ad.X.<span class="built_in">max</span>()</span><br></pre></td></tr></table></figure>
<pre><code>1137.2114
</code></pre>
<p>We recommend that the data be log transformed. Note that, some datasets show better signal in the linear scale while others show stronger signal in the log scale.个人认为如果没有特殊情况，log即可</p>
<p>palantir的这个<code>log_transform</code>不同于 <code>sc.pp.log1p</code>是不同的，但是差别不大，因为只需要修改pseudo_count&#x3D;1，以及我们的底数不太一样，但对结果影响不大</p>
<p>log1p: Logarithmize the data matrix.Computes $(X &#x3D; log(X + 1))$, where (log) denotes the natural logarithm unless a different base is given.</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">def log_transform(data, pseudo_count=0.1):</span><br><span class="line">    &quot;&quot;&quot;Log transform the matrix</span><br><span class="line"></span><br><span class="line">    :param data: Counts matrix: Cells x Genes or Anndata object</span><br><span class="line">    :return: Log transformed matrix</span><br><span class="line">    &quot;&quot;&quot;</span><br><span class="line">    if isinstance(data, anndata.AnnData):</span><br><span class="line">        if issparse(data.X):</span><br><span class="line">            data.X.data = np.log2(data.X.data + pseudo_count) - np.log2(pseudo_count)</span><br><span class="line">        else:</span><br><span class="line">            data.X = np.log2(data.X + pseudo_count) - np.log2(pseudo_count)</span><br><span class="line">    else:</span><br><span class="line">        return np.log2(data + pseudo_count)</span><br></pre></td></tr></table></figure>


<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pa.preprocess.log_transform(ad)</span><br><span class="line">ad.X.<span class="built_in">max</span>()</span><br></pre></td></tr></table></figure>
<pre><code>13.47334
</code></pre>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Highly variable gene selection</span></span><br><span class="line"><span class="comment"># Highly variable gene selection can also be performed using the scanpy interface</span></span><br><span class="line">sc.pp.highly_variable_genes(ad, n_top_genes=<span class="number">2000</span>, flavor=<span class="string">&quot;cell_ranger&quot;</span>)</span><br></pre></td></tr></table></figure>
<ul>
<li>这儿的<code>falavor</code>参数注意一下：<strong>flavor</strong> : <a target="_blank" rel="noopener" href="https://docs.python.org/3/library/typing.html#typing.Literal" title="(in Python v3.11)"><code>Literal</code></a>[<code>&#39;seurat&#39;</code>, <code>&#39;cell_ranger&#39;</code>, <code>&#39;seurat_v3&#39;</code>] (default: <code>&#39;seurat&#39;</code>).Choose the flavor for identifying highly variable genes. For the dispersion based methods in their default workflows, Seurat passes the cutoffs whereas Cell Ranger passes <code>n_top_genes</code>.</li>
</ul>
<p>PCA is the first step in data processing for Palantir. This representation is necessary to overcome the extensive dropouts that are pervasive in single cell RNA-seq data. pca是我们降噪必然会做的一步</p>
<p>Rather than use a fixed number of PCs, we recommend the use of components that explain 85% of the variance in the data after highly variable gene selection.</p>
<p>这儿我们直接使用default_pca&#x3D;50</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Note in the manuscript, we did not use highly variable genes </span></span><br><span class="line"><span class="comment"># but scanpy by default uses only highly variable genes</span></span><br><span class="line"><span class="comment"># Computes PCA coordinates, loadings and variance decomposition. Uses the implementation of scikit-learn</span></span><br><span class="line">sc.pp.pca(ad)</span><br><span class="line">ad</span><br></pre></td></tr></table></figure>
<pre><code>AnnData object with n_obs × n_vars = 4142 × 16106
    var: &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;hvg&#39;, &#39;pca&#39;
    obsm: &#39;X_pca&#39;
    varm: &#39;PCs&#39;
</code></pre>
<h1 id="2-Diffusion-maps"><a href="#2-Diffusion-maps" class="headerlink" title="2. Diffusion maps"></a>2. Diffusion maps</h1><p>Palantir next determines the diffusion maps of the data as an estimate of the low dimensional phenotypic manifold of the data.</p>
<p>我们看一下源代码</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">def run_diffusion_maps(</span><br><span class="line">    data: Union[pd.DataFrame, sc.AnnData],</span><br><span class="line">    n_components: int = 10,</span><br><span class="line">    knn: int = 30,</span><br><span class="line">    alpha: float = 0,</span><br><span class="line">    seed: Union[int, None] = 0,</span><br><span class="line">    pca_key: str = &quot;X_pca&quot;,</span><br><span class="line">    kernel_key: str = &quot;DM_Kernel&quot;,</span><br><span class="line">    sim_key: str = &quot;DM_Similarity&quot;,</span><br><span class="line">    eigval_key: str = &quot;DM_EigenValues&quot;,</span><br><span class="line">    eigvec_key: str = &quot;DM_EigenVectors&quot;,</span><br><span class="line">)</span><br><span class="line"># ....</span><br><span class="line">    kernel = compute_kernel(data_df, knn, alpha)</span><br><span class="line"># ....</span><br><span class="line">    res = diffusion_maps_from_kernel(kernel, n_components, seed)</span><br><span class="line"></span><br><span class="line">    res[&quot;kernel&quot;] = kernel</span><br><span class="line"></span><br><span class="line">    if isinstance(data, sc.AnnData):</span><br><span class="line">        data.obsp[kernel_key] = res[&quot;kernel&quot;]</span><br><span class="line">        data.obsp[sim_key] = res[&quot;T&quot;]</span><br><span class="line">        data.obsm[eigvec_key] = res[&quot;EigenVectors&quot;].values</span><br><span class="line">        data.uns[eigval_key] = res[&quot;EigenValues&quot;].values</span><br><span class="line"></span><br><span class="line">    return res</span><br></pre></td></tr></table></figure>

<p><code>compute_kernel</code> 计算了neighbor和knn</p>
<p>核心的算法是在<code>diffusion_maps_from_kernel</code>, 主要干的事情是：执行了一些矩阵运算和特征值分解</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Run diffusion maps</span></span><br><span class="line">dm_res = pa.utils.run_diffusion_maps(ad, n_components=<span class="number">5</span>)</span><br></pre></td></tr></table></figure>

<p>The low dimensional embeddeing of the data is estimated based on the eigen gap using the following function</p>
<p>我们观察源码发现，主要就是缩放并提取特征向量</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">def determine_multiscale_space(</span><br><span class="line">#....</span><br><span class="line"># 缩放数据</span><br><span class="line"></span><br><span class="line"># 从之前的计算结果中获取特征值（EigenValues）和特征向量（EigenVectors）</span><br><span class="line"># use_eigs是一个包含要使用的特征值索引的列表，从1到n_eigs-1</span><br><span class="line">use_eigs = list(range(1, n_eigs))</span><br><span class="line">eig_vals = np.ravel(dm_res_dict[&quot;EigenValues&quot;][use_eigs])</span><br><span class="line"></span><br><span class="line"># 从特征向量中提取需要的列，并进行缩放操作</span><br><span class="line"># 缩放操作涉及特征值（eig_vals）的计算，然后用特征向量的对应列乘以缩放因子</span><br><span class="line">data = dm_res_dict[&quot;EigenVectors&quot;].values[:, use_eigs] * (eig_vals / (1 - eig_vals))</span><br><span class="line"></span><br><span class="line"># 将缩放后的数据存储到一个新的DataFrame中，以及设置相应的索引</span><br><span class="line">data = pd.DataFrame(data, index=dm_res_dict[&quot;EigenVectors&quot;].index)</span><br><span class="line"></span><br><span class="line"># 检查dm_res是否是AnnData对象</span><br><span class="line">if isinstance(dm_res, sc.AnnData):</span><br><span class="line">    # 如果是AnnData对象，则将缩放后的数据存储到obsm属性中的指定键（out_key）</span><br><span class="line">    dm_res.obsm[out_key] = data.values</span><br><span class="line"></span><br><span class="line">)</span><br></pre></td></tr></table></figure>


<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">ms_data = pa.utils.determine_multiscale_space(ad,n_eigs=<span class="number">5</span>)</span><br><span class="line">ms_data</span><br></pre></td></tr></table></figure>

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

<pre><code>.dataframe tbody tr th &#123;
    vertical-align: top;
&#125;

.dataframe thead th &#123;
    text-align: right;
&#125;
</code></pre>
<p></style><p></p>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Run4_120703408880541</th>
      <td>-3.764623</td>
      <td>-0.244870</td>
      <td>-0.003447</td>
      <td>0.338470</td>
    </tr>
    <tr>
      <th>Run4_120703409056541</th>
      <td>0.187935</td>
      <td>0.988136</td>
      <td>0.074741</td>
      <td>0.078343</td>
    </tr>
    <tr>
      <th>Run4_120703409580963</th>
      <td>1.754346</td>
      <td>-1.233746</td>
      <td>0.356473</td>
      <td>0.345751</td>
    </tr>
    <tr>
      <th>Run4_120703423990708</th>
      <td>1.682744</td>
      <td>-1.139355</td>
      <td>0.512876</td>
      <td>0.082349</td>
    </tr>
    <tr>
      <th>Run4_120703436876077</th>
      <td>0.282202</td>
      <td>0.877546</td>
      <td>0.062346</td>
      <td>0.015017</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Run5_241098904976174</th>
      <td>0.763714</td>
      <td>0.451856</td>
      <td>-0.161372</td>
      <td>-0.176209</td>
    </tr>
    <tr>
      <th>Run5_241106375007076</th>
      <td>-4.872839</td>
      <td>-0.658010</td>
      <td>-0.050222</td>
      <td>-0.184598</td>
    </tr>
    <tr>
      <th>Run5_241114577000174</th>
      <td>0.605820</td>
      <td>0.592617</td>
      <td>0.020493</td>
      <td>-0.127850</td>
    </tr>
    <tr>
      <th>Run5_241114577004764</th>
      <td>1.178952</td>
      <td>-0.152643</td>
      <td>-0.370034</td>
      <td>-0.090144</td>
    </tr>
    <tr>
      <th>Run5_241114589051630</th>
      <td>-5.269270</td>
      <td>-0.815916</td>
      <td>-0.070615</td>
      <td>-0.520574</td>
    </tr>
  </tbody>
</table>
<p>4142 rows × 4 columns</p>
</div>


<p>If you are specifying the number of eigen vectors manually in the above step, please ensure that the specified parameter is &gt; 2，当然你不输入<code>n_eigs</code>，palantir会自动帮你赋值，并且最终的结果可以看到列数为$n_{eigs}-1$</p>
<h1 id="3-Visualization"><a href="#3-Visualization" class="headerlink" title="3. Visualization"></a>3. Visualization</h1><p>In the manuscript, we used tSNE projection using diffusion components to visualize the data. We now recommend the use of force-directed layouts for visualization of trajectories. Force-directed layouts can be computed by the same adaptive kernel used for determining diffusion maps.</p>
<p>scanpy can be used to compute force directed layouts. We recommened the use of the diffusion kernel (see below) for computing force directed layouts Force Atlas package should be installed for this analysis and can be installed using conda install -c conda-forge fa2. Note that fa2 is not supported by python3.9<br>这里要特别注意的是，我们使用scvi-tools往往是基于3.9的情况，所以如果需要使用这个功能，请使用3.8版本的python</p>
<p>UMAPs are a good alternative to visualize trajectories in addition to force directed layouts.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># use the default paramters</span></span><br><span class="line">sc.pp.neighbors(ad)</span><br><span class="line">sc.tl.umap(ad)</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Use scanpy functions to visualize umaps or FDL</span></span><br><span class="line">sc.pl.embedding(</span><br><span class="line">    ad,</span><br><span class="line">    basis=<span class="string">&quot;umap&quot;</span>,</span><br><span class="line">    frameon=<span class="literal">True</span>,</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_26_0.png" alt="png"></p>
<h1 id="4-MAGIC-imputation"><a href="#4-MAGIC-imputation" class="headerlink" title="4. MAGIC imputation"></a>4. MAGIC imputation</h1><p>MAGIC is an imputation technique developed in the Pe’er lab for single cell data imputation. Palantir uses MAGIC to impute the data for visualization and determining gene expression trends.</p>
<p><strong>Markov Affinity-based Graph Imputation of Cells (MAGIC)</strong> is an algorithm for denoising high-dimensional data most commonly applied to single-cell RNA sequencing data. MAGIC learns the manifold data, using the resultant graph to smooth the features and restore the structure of the data.</p>
<ul>
<li>这是一种基于马尔可夫亲和的细胞图算法，可以很好地去噪点（降维）<br><img src="/2023/09/18/pigudog/pigudog1-palantir/magic.gif"></li>
</ul>
<p>Palantir使用MAGIC来计算可视化数据和确定基因表达趋势。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">imputed_X = pa.utils.run_magic_imputation(ad,n_jobs=<span class="number">16</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">imputed_X.shape</span><br></pre></td></tr></table></figure>
<pre><code>(4142, 16106)
</code></pre>
<p>Gene expression can be visualized on umaps using the scanpy functions. The genes parameter is an string iterable of genes, which are a subset of the expression of column names. The below function plots the expression of HSC gene CD34, myeloid gene MPO and erythroid precursor gene GATA1 and dendritic cell gene IRF8.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">sc.pl.embedding(</span><br><span class="line">    ad,</span><br><span class="line">    basis=<span class="string">&quot;umap&quot;</span>,</span><br><span class="line">    layer=<span class="string">&quot;MAGIC_imputed_data&quot;</span>,</span><br><span class="line">    color=[<span class="string">&quot;CD34&quot;</span>, <span class="string">&quot;MPO&quot;</span>, <span class="string">&quot;GATA1&quot;</span>, <span class="string">&quot;IRF8&quot;</span>],</span><br><span class="line">    frameon=<span class="literal">False</span>,</span><br><span class="line">    cmap=<span class="string">&#x27;PuRd&#x27;</span></span><br><span class="line">)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_31_0.png" alt="png"></p>
<h1 id="5-Diffusion-maps-visualization"><a href="#5-Diffusion-maps-visualization" class="headerlink" title="5. Diffusion maps visualization"></a>5. Diffusion maps visualization</h1><p>The computed diffusion components can be visualized with the following snippet.<br>可以看到我们需要的是component 2</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pa.plot.plot_diffusion_components(ad)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_33_0.png" alt="png"></p>
<h1 id="6-Running-Palantir"><a href="#6-Running-Palantir" class="headerlink" title="6. Running Palantir"></a>6. Running Palantir</h1><p>Palantir can be run by specifying an approxiate early cell.</p>
<p>Palantir can automatically determine the terminal states as well. In this dataset, we know the terminal states and we will set them using the terminal_states parameter</p>
<p>The start cell for this dataset was chosen based on high expression of CD34.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">terminal_states = pd.Series(</span><br><span class="line">    [<span class="string">&quot;DC&quot;</span>, <span class="string">&quot;Mono&quot;</span>, <span class="string">&quot;Ery&quot;</span>],</span><br><span class="line">    index=[<span class="string">&quot;Run5_131097901611291&quot;</span>, <span class="string">&quot;Run5_134936662236454&quot;</span>, <span class="string">&quot;Run4_200562869397916&quot;</span>],</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pa.plot.highlight_cells_on_umap(ad, terminal_states)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_36_0.png" alt="png"></p>
<p>我们查看一下最核心的代码<code>run_palantir</code></p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br></pre></td><td class="code"><pre><span class="line">def run_palantir(</span><br><span class="line">    # 使用了Python中的Union类型注解，用于表示一个变量可以接受两种不同类型的值</span><br><span class="line">    data: Union[pd.DataFrame, sc.AnnData],</span><br><span class="line">    early_cell,</span><br><span class="line">    # &quot;Optional&quot;是一个类型注解，表示这个变量可以是可选的，也就是可以不提供值。如果不提供值，它将默认为&quot;None&quot;。</span><br><span class="line">    terminal_states: Optional[Union[List, Dict, pd.Series]] = None,</span><br><span class="line">    knn: int = 30,</span><br><span class="line">    num_waypoints: int = 1200,</span><br><span class="line">    n_jobs: int = -1,</span><br><span class="line">    scale_components: bool = True,</span><br><span class="line">    use_early_cell_as_start: bool = False,</span><br><span class="line">    max_iterations: int = 25,</span><br><span class="line">    eigvec_key: str = &quot;DM_EigenVectors_multiscaled&quot;,</span><br><span class="line">    pseudo_time_key: str = &quot;palantir_pseudotime&quot;,</span><br><span class="line">    entropy_key: str = &quot;palantir_entropy&quot;,</span><br><span class="line">    fate_prob_key: str = &quot;palantir_fate_probabilities&quot;,</span><br><span class="line">    save_as_df: bool = None,</span><br><span class="line">    waypoints_key: str = &quot;palantir_waypoints&quot;,</span><br><span class="line">    seed: int = 20, # 自动设置的seed</span><br><span class="line">) -&gt; Optional[PResults]:</span><br><span class="line">    &quot;&quot;&quot;</span><br><span class="line">    Executes the Palantir algorithm to derive pseudotemporal ordering of cells, their fate probabilities, and</span><br><span class="line">    state entropy based on the multiscale diffusion map results.</span><br><span class="line"></span><br><span class="line">    Parameters</span><br><span class="line">    ----------</span><br><span class="line">    data : Union[pd.DataFrame, sc.AnnData]</span><br><span class="line">        Either a DataFrame of multiscale space diffusion components or a Scanpy AnnData object.</span><br><span class="line">    early_cell : str</span><br><span class="line">        Start cell for pseudotime construction.</span><br><span class="line">    terminal_states : List/Series/Dict, optional</span><br><span class="line">        User-defined terminal states structure in the format &#123;terminal_name:cell_name&#125;. Default is None.</span><br><span class="line">    knn : int, optional</span><br><span class="line">        Number of nearest neighbors for graph construction. Default is 30.</span><br><span class="line">    num_waypoints : int, optional</span><br><span class="line">        Number of waypoints to sample. Default is 1200.</span><br><span class="line">    n_jobs : int, optional</span><br><span class="line">        Number of jobs for parallel processing. Default is -1.</span><br><span class="line">    scale_components : bool, optional</span><br><span class="line">        If True, components are scaled. Default is True.</span><br><span class="line">    use_early_cell_as_start : bool, optional</span><br><span class="line">        If True, the early cell is used as start. Default is False.</span><br><span class="line">    max_iterations : int, optional</span><br><span class="line">        Maximum number of iterations for pseudotime convergence. Default is 25.</span><br><span class="line">    eigvec_key : str, optional</span><br><span class="line">        Key to access multiscale space diffusion components from obsm of the AnnData object. Default is &#x27;DM_EigenVectors_multiscaled&#x27;.</span><br><span class="line">    pseudo_time_key : str, optional</span><br><span class="line">        Key to store the pseudotime in obs of the AnnData object. Default is &#x27;palantir_pseudotime&#x27;.</span><br><span class="line">    entropy_key : str, optional</span><br><span class="line">        Key to store the entropy in obs of the AnnData object. Default is &#x27;palantir_entropy&#x27;.</span><br><span class="line">    fate_prob_key : str, optional</span><br><span class="line">        Key to store the fate probabilities in obsm of the AnnData object. Default is &#x27;palantir_fate_probabilities&#x27;.</span><br><span class="line">        If save_as_df is True, the fate probabilities are stored as pandas DataFrame with terminal state names as columns.</span><br><span class="line">        If False, the fate probabilities are stored as numpy array and the terminal state names are stored in uns[fate_prob_key + &quot;_columns&quot;].</span><br><span class="line">    save_as_df : bool, optional</span><br><span class="line">        If True, the fate probabilities are saved as pandas DataFrame. If False, the data is saved as numpy array.</span><br><span class="line">        The option to save as DataFrame is there due to some versions of AnnData not being able to</span><br><span class="line">        write h5ad files with DataFrames in ad.obsm. Default is palantir.SAVE_AS_DF = True.</span><br><span class="line">    waypoints_key : str, optional</span><br><span class="line">        Key to store the waypoints in uns of the AnnData object. Default is &#x27;palantir_waypoints&#x27;.</span><br><span class="line">    seed : int, optional</span><br><span class="line">        The seed for the random number generator used in waypoint sampling. Default is 20.</span><br><span class="line"></span><br><span class="line">    Returns</span><br><span class="line">    -------</span><br><span class="line">    Optional[PResults]</span><br><span class="line">        PResults object with pseudotime, entropy, branch probabilities, and waypoints.</span><br><span class="line">        If an AnnData object is passed as data, the result is written to its obs, obsm, and uns attributes</span><br><span class="line">        using the provided keys and None is returned.</span><br><span class="line">    &quot;&quot;&quot;</span><br><span class="line">    # the default of SAVE_AS_DF = True</span><br><span class="line">    if save_as_df is None:</span><br><span class="line">        save_as_df = config.SAVE_AS_DF</span><br><span class="line"></span><br><span class="line">    # 对应于不同的类型的terminal_states</span><br><span class="line">    # List/Series/Dict</span><br><span class="line">    # 最终都转化为List</span><br><span class="line">    if isinstance(terminal_states, dict):</span><br><span class="line">        terminal_states = pd.Series(terminal_states)</span><br><span class="line">    if isinstance(terminal_states, pd.Series):</span><br><span class="line">        terminal_cells = terminal_states.index.values</span><br><span class="line">    else:</span><br><span class="line">        terminal_cells = terminal_states</span><br><span class="line">        </span><br><span class="line">    # 对应于输入的是datafram还是anndata</span><br><span class="line">    # Key to access multiscale space diffusion components from obsm of the AnnData object. Default is &#x27;DM_EigenVectors_multiscaled&#x27;.</span><br><span class="line">    if isinstance(data, sc.AnnData):</span><br><span class="line">        ms_data = pd.DataFrame(data.obsm[eigvec_key], index=data.obs_names)</span><br><span class="line">    else:</span><br><span class="line">        ms_data = data</span><br><span class="line"></span><br><span class="line">    # 如果&quot;scale_components&quot;为真，就对数据进行最小-最大缩放，并将结果保存在&quot;data_df&quot;中。</span><br><span class="line">    # 如果&quot;scale_components&quot;为假，就创建&quot;data_df&quot;作为&quot;ms_data&quot;的副本，不进行缩放。</span><br><span class="line">    if scale_components:</span><br><span class="line">        data_df = pd.DataFrame(</span><br><span class="line">            preprocessing.minmax_scale(ms_data),</span><br><span class="line">            index=ms_data.index,</span><br><span class="line">            columns=ms_data.columns,</span><br><span class="line">        )</span><br><span class="line">    else:</span><br><span class="line">        data_df = copy.copy(ms_data)</span><br><span class="line"></span><br><span class="line">    # ################################################</span><br><span class="line">    # Determine the boundary cell closest to user defined early cell</span><br><span class="line">    # 这段代码的目的是找到与用户指定的&quot;early_cell&quot;最接近的边界细胞，并根据&quot;use_early_cell_as_start&quot;标志决定最终的起始细胞是&quot;early_cell&quot;还是最接近的边界细胞。</span><br><span class="line">    dm_boundaries = pd.Index(set(data_df.idxmax()).union(data_df.idxmin()))</span><br><span class="line">    dists = pairwise_distances(</span><br><span class="line">        data_df.loc[dm_boundaries, :], data_df.loc[early_cell, :].values.reshape(1, -1)</span><br><span class="line">    )</span><br><span class="line">    start_cell = pd.Series(np.ravel(dists), index=dm_boundaries).idxmin()</span><br><span class="line">    if use_early_cell_as_start:</span><br><span class="line">        start_cell = early_cell</span><br><span class="line">        </span><br><span class="line">        </span><br><span class="line">    # 综合起来，这段代码的作用是创建一组称为&quot;waypoints&quot;的细胞，其中包括了起始细胞、边界细胞、以及可能的终止细胞。</span><br><span class="line">    # 这个&quot;waypoints&quot;集合将用于后续的操作，可能是路径规划或其他任务</span><br><span class="line">    # Sample waypoints</span><br><span class="line">    print(&quot;Sampling and flocking waypoints...&quot;)</span><br><span class="line">    start = time.time()</span><br><span class="line">    # Append start cell</span><br><span class="line">    if isinstance(num_waypoints, int):</span><br><span class="line">        waypoints = _max_min_sampling(data_df, num_waypoints, seed)</span><br><span class="line">    else:</span><br><span class="line">        waypoints = num_waypoints</span><br><span class="line">    waypoints = waypoints.union(dm_boundaries)</span><br><span class="line">    if terminal_cells is not None:</span><br><span class="line">        waypoints = waypoints.union(terminal_cells)</span><br><span class="line">    waypoints = pd.Index(waypoints.difference([start_cell]).unique())</span><br><span class="line">    # Append start cell</span><br><span class="line">    waypoints = pd.Index([start_cell]).append(waypoints)</span><br><span class="line">    end = time.time()</span><br><span class="line">    print(&quot;Time for determining waypoints: &#123;&#125; minutes&quot;.format((end - start) / 60))</span><br><span class="line">    </span><br><span class="line">    # 这段代码的主要目的是对细胞数据进行分析，</span><br><span class="line">    # 包括伪时间计算、熵计算以及投影结果的计算，并将结果存储在相应的变量中。</span><br><span class="line">    # 这些计算通常用于分析单细胞RNA测序数据中的细胞状态和分支情况</span><br><span class="line">    # pseudotime and weighting matrix</span><br><span class="line">    print(&quot;Determining pseudotime...&quot;)</span><br><span class="line">    pseudotime, W = _compute_pseudotime(</span><br><span class="line">        data_df, start_cell, knn, waypoints, n_jobs, max_iterations</span><br><span class="line">    )</span><br><span class="line"></span><br><span class="line">    # Entropy and branch probabilities</span><br><span class="line">    print(&quot;Entropy and branch probabilities...&quot;)</span><br><span class="line">    ent, branch_probs = _differentiation_entropy(</span><br><span class="line">        data_df.loc[waypoints, :], terminal_cells, knn, n_jobs, pseudotime</span><br><span class="line">    )</span><br><span class="line"></span><br><span class="line">    # Project results to all cells</span><br><span class="line">    print(&quot;Project results to all cells...&quot;)</span><br><span class="line">    branch_probs = pd.DataFrame(</span><br><span class="line">        np.dot(W.T, branch_probs.loc[W.index, :]),</span><br><span class="line">        index=W.columns,</span><br><span class="line">        columns=branch_probs.columns,</span><br><span class="line">    )</span><br><span class="line">    ent = branch_probs.apply(entropy, axis=1)</span><br><span class="line"></span><br><span class="line">    # Update results into PResults class object</span><br><span class="line">    pr_res = PResults(pseudotime, ent, branch_probs, waypoints)</span><br><span class="line"></span><br><span class="line">    if isinstance(data, sc.AnnData):</span><br><span class="line">        data.obs[pseudo_time_key] = pseudotime</span><br><span class="line">        data.obs[entropy_key] = ent</span><br><span class="line">        data.uns[waypoints_key] = waypoints.values</span><br><span class="line">        if isinstance(terminal_states, pd.Series):</span><br><span class="line">            branch_probs.columns = terminal_states[branch_probs.columns]</span><br><span class="line">        if save_as_df:</span><br><span class="line">            data.obsm[fate_prob_key] = branch_probs</span><br><span class="line">        else:</span><br><span class="line">            data.obsm[fate_prob_key] = branch_probs.values</span><br><span class="line">            data.uns[fate_prob_key + &quot;_columns&quot;] = branch_probs.columns.values</span><br><span class="line"></span><br><span class="line">    return pr_res</span><br></pre></td></tr></table></figure>


<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">start_cell = <span class="string">&quot;Run5_164698952452459&quot;</span></span><br><span class="line">pr_res = pa.core.run_palantir(</span><br><span class="line">    ad, start_cell, num_waypoints=<span class="number">500</span>, terminal_states=terminal_states,</span><br><span class="line">    n_jobs=<span class="number">16</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<pre><code>Sampling and flocking waypoints...
Time for determining waypoints: 0.01345531145731608 minutes
Determining pseudotime...
Shortest path distances using 30-nearest neighbor graph...
Time for shortest paths: 1.9907963196436564 minutes
Iteratively refining the pseudotime...
Correlation at iteration 1: 0.9999
Entropy and branch probabilities...
Markov chain construction...
Computing fundamental matrix and absorption probabilities...
Project results to all cells...
</code></pre>
<p>Palantir generates the following results</p>
<pre><code>Pseudotime: Pseudo time ordering of each cell
Terminal state probabilities: Matrix of cells X terminal states. Each entry represents the probability of the corresponding cell reaching the respective terminal state
Entropy: A quantiative measure of the differentiation potential of each cell computed as the entropy of the multinomial terminal state probabilities
</code></pre>
<h1 id="7-Visualizing-Palantir-results"><a href="#7-Visualizing-Palantir-results" class="headerlink" title="7. Visualizing Palantir results"></a>7. Visualizing Palantir results</h1><p>Palantir results can be visualized on the tSNE or UMAP using the plot_palantir_results function</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pa.plot.plot_palantir_results(ad, s=<span class="number">10</span>)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_41_0.png" alt="png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Terminal state probability distributions of individual cells can be visualized using the plot_terminal_state_probs function</span></span><br><span class="line"></span><br><span class="line">cells = [</span><br><span class="line">    <span class="string">&quot;Run5_164698952452459&quot;</span>,</span><br><span class="line">    <span class="string">&quot;Run5_170327461775790&quot;</span>,</span><br><span class="line">    <span class="string">&quot;Run4_121896095574750&quot;</span>,</span><br><span class="line">]</span><br><span class="line">pa.plot.plot_terminal_state_probs(ad, cells)</span><br><span class="line">plt.show()</span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_42_0.png" alt="png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pa.plot.highlight_cells_on_umap(ad, cells)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_43_0.png" alt="png"></p>
<h1 id="8-Gene-expression-trends"><a href="#8-Gene-expression-trends" class="headerlink" title="8. Gene expression trends"></a>8. Gene expression trends</h1><p>Gene expression trends over pseudotime provide insights into the dynamic behavior of genes during cellular development or progression. By examining these trends, we can uncover the timing of gene expression changes and identify pivotal regulators of cellular states. Palantir provides tools for computing these gene expression trends.</p>
<p>Here, we’ll outline the steps to compute gene trends over pseudotime using Palantir.</p>
<p><strong>Selecting cells of a specific trend</strong></p>
<p>Before computing the gene expression trends, we first need to select cells associated with a specific branch of the pseudotime trajectory. We accomplish this by using the select_branch_cells function. The parameter q is used to control the selection’s stringency.</p>
<p>我们查阅源码发现每个branch的选取：</p>
<ul>
<li>我们通过之前得到的每个分化的pseduotime去取quantile得到<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">masks = pa.presults.select_branch_cells(ad, eps=<span class="number">0</span>)</span><br><span class="line">masks</span><br></pre></td></tr></table></figure>
  array([[ True, False, False],<br>     [False, False, False],<br>     [False, False, False],<br>     …,<br>     [False,  True,  True],<br>     [False, False, False],<br>     [ True, False, False]])</li>
</ul>
<p><strong>Visualizing the branch selection</strong><br>Once the cells are selected, it’s often helpful to visualize the selection on the pseudotime trajectory to ensure we’ve isolated the correct cells for our specific trend. We can do this using the plot_branch_selection function:</p>
<p>可以看到分化结果非常好</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">pa.plot.plot_branch_selection(ad)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_47_0.png" alt="png"></p>
<p>Palantir uses <strong>Mellon</strong> Function Estimator to determine the gene expression trends along different lineages. The marker trends can be determined using the following snippet. This computes the trends for all lineages. A subset of lineages can be used using the lineages parameter.</p>
<p>我们通过一下的代码查看我们的基因趋势</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">gene_trends = pa.presults.compute_gene_trends(</span><br><span class="line">    ad,</span><br><span class="line">    expression_key=<span class="string">&quot;MAGIC_imputed_data&quot;</span>,</span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<pre><code>Ery
[2023-09-18 16:34:16,254] [INFO    ] Using covariance function Matern52(ls=1.0).
DC
[2023-09-18 16:34:28,307] [INFO    ] Using covariance function Matern52(ls=1.0).
Mono
[2023-09-18 16:34:30,480] [INFO    ] Using covariance function Matern52(ls=1.0).
</code></pre>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">genes = [<span class="string">&quot;CD34&quot;</span>, <span class="string">&quot;MPO&quot;</span>, <span class="string">&quot;GATA1&quot;</span>, <span class="string">&quot;IRF8&quot;</span>]</span><br><span class="line">pa.plot.plot_gene_trends(ad, genes)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>

<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_50_0.png" alt="png"></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Alternatively, the trends can be visualized on a heatmap using</span></span><br><span class="line">pa.plot.plot_gene_trend_heatmaps(ad, genes)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_51_0.png" alt="png"></p>
<p> <strong>Clustering</strong></p>
<p>Gene expression trends can be clustered and visualized using the following snippet. As an example, the first 1000 genes along the erythroid genes are clustered</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">more_genes = ad.var_names[:<span class="number">1000</span>]</span><br><span class="line">communities = pa.presults.cluster_gene_trends(ad, <span class="string">&quot;Ery&quot;</span>, more_genes)</span><br><span class="line"></span><br><span class="line">pa.plot.plot_gene_trend_clusters(ad, <span class="string">&quot;Ery&quot;</span>)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_53_0.png" alt="png"></p>
<p>比如我们找这个cluster0，因为他似乎是全部为上升趋势的</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gene_cluster0 = ad.var_names[ad.var.gene_trends_clusters==<span class="string">&#x27;0&#x27;</span>]</span><br></pre></td></tr></table></figure>

<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gene_cluster0</span><br></pre></td></tr></table></figure>
<pre><code>Index([&#39;FHL2&#39;, &#39;FUT10&#39;, &#39;PIGX&#39;, &#39;OLA1&#39;, &#39;RBM6&#39;, &#39;RP11-262H14.3&#39;, &#39;COX17&#39;,
       &#39;OSGEP&#39;, &#39;GPR137&#39;, &#39;CMIP&#39;,
       ...
       &#39;ANAPC1&#39;, &#39;CASD1&#39;, &#39;CTDSPL2&#39;, &#39;SMPD1&#39;, &#39;POLR2G&#39;, &#39;TM7SF3&#39;, &#39;OMA1&#39;,
       &#39;ALDH6A1&#39;, &#39;TOLLIP&#39;, &#39;XRCC6&#39;],
      dtype=&#39;object&#39;, length=156)
</code></pre>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">sc.pl.embedding(</span><br><span class="line">    ad,</span><br><span class="line">    basis=<span class="string">&quot;umap&quot;</span>,</span><br><span class="line">    layer=<span class="string">&quot;MAGIC_imputed_data&quot;</span>,</span><br><span class="line">    color=[<span class="string">&#x27;FHL2&#x27;</span>, <span class="string">&#x27;FUT10&#x27;</span>, <span class="string">&#x27;PIGX&#x27;</span>],</span><br><span class="line">    frameon=<span class="literal">False</span>,</span><br><span class="line">    cmap=<span class="string">&#x27;PuRd&#x27;</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p><img src="/2023/09/18/pigudog/pigudog1-palantir/step7_palantir_modify_56_0.png" alt="png"></p>
<p><strong>Save</strong></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Save results</span></span><br><span class="line">file_path = os.path.join(data_dir, <span class="string">&quot;marrow_sample_scseq_processed.h5ad&quot;</span>)</span><br><span class="line">ad.write(file_path)</span><br></pre></td></tr></table></figure>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="http://example.com">pigudog</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://example.com/2023/09/18/pigudog/pigudog1-palantir/">http://example.com/2023/09/18/pigudog/pigudog1-palantir/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"></div><div class="post_share"><div class="social-share" data-image="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><div class="post-reward"><div class="reward-button"><i class="fas fa-qrcode"></i> Donate</div><div class="reward-main"><ul class="reward-all"><li class="reward-item"><a href="/img/wechat.jpg" target="_blank"><img class="post-qr-code-img" src="/img/wechat.jpg" alt="wechat"/></a><div class="post-qr-code-desc">wechat</div></li></ul></div></div><nav class="pagination-post" id="pagination"><div class="next-post pull-full"><a href="/2023/02/16/statistic/statistic7%20-%20Analysis%20of%20variance%20and%20regression/"><img class="next-cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">statistic7 - Analysis of variance and regression</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/img/avatar.gif" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">pigudog</div><div class="author-info__description">冲冲冲！</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">69</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">12</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">16</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/pigudog"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/pigudog" target="_blank" title="Github"><i class="fab fa-github"></i></a><a class="social-icon" href="/2205638116@qq.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#Introduction"><span class="toc-number">1.</span> <span class="toc-text">Introduction</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Markov-chain"><span class="toc-number">1.1.</span> <span class="toc-text">Markov chain</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%9F%A5%E7%9C%8B%E6%BA%90%E7%A0%81"><span class="toc-number">1.2.</span> <span class="toc-text">查看源码</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#preparatory-work"><span class="toc-number">2.</span> <span class="toc-text">preparatory work</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#1-load-and-preprocess"><span class="toc-number">3.</span> <span class="toc-text">1. load and preprocess</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#2-preprocess"><span class="toc-number">4.</span> <span class="toc-text">2. preprocess</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#2-Diffusion-maps"><span class="toc-number">5.</span> <span class="toc-text">2. Diffusion maps</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#3-Visualization"><span class="toc-number">6.</span> <span class="toc-text">3. Visualization</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#4-MAGIC-imputation"><span class="toc-number">7.</span> <span class="toc-text">4. MAGIC imputation</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#5-Diffusion-maps-visualization"><span class="toc-number">8.</span> <span class="toc-text">5. Diffusion maps visualization</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#6-Running-Palantir"><span class="toc-number">9.</span> <span class="toc-text">6. Running Palantir</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#7-Visualizing-Palantir-results"><span class="toc-number">10.</span> <span class="toc-text">7. Visualizing Palantir results</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#8-Gene-expression-trends"><span class="toc-number">11.</span> <span class="toc-text">8. Gene expression trends</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2023/09/18/pigudog/pigudog1-palantir/" title="pigudog1-palantir"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="pigudog1-palantir"/></a><div class="content"><a class="title" href="/2023/09/18/pigudog/pigudog1-palantir/" title="pigudog1-palantir">pigudog1-palantir</a><time datetime="2023-09-17T16:00:00.000Z" title="Created 2023-09-18 00:00:00">2023-09-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/16/statistic/statistic7%20-%20Analysis%20of%20variance%20and%20regression/" title="statistic7 - Analysis of variance and regression"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic7 - Analysis of variance and regression"/></a><div class="content"><a class="title" href="/2023/02/16/statistic/statistic7%20-%20Analysis%20of%20variance%20and%20regression/" title="statistic7 - Analysis of variance and regression">statistic7 - Analysis of variance and regression</a><time datetime="2023-02-15T16:00:00.000Z" title="Created 2023-02-16 00:00:00">2023-02-16</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/15/statistic/statistic6%20-%20Hypothesis%20testing/" title="statistic6 - Hypothesis testing"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic6 - Hypothesis testing"/></a><div class="content"><a class="title" href="/2023/02/15/statistic/statistic6%20-%20Hypothesis%20testing/" title="statistic6 - Hypothesis testing">statistic6 - Hypothesis testing</a><time datetime="2023-02-14T16:00:00.000Z" title="Created 2023-02-15 00:00:00">2023-02-15</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/13/statistic/statistic4%20-%20Sampling/" title="statistic4 - Sampling"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic4 - Sampling"/></a><div class="content"><a class="title" href="/2023/02/13/statistic/statistic4%20-%20Sampling/" title="statistic4 - Sampling">statistic4 - Sampling</a><time datetime="2023-02-12T16:00:00.000Z" title="Created 2023-02-13 00:00:00">2023-02-13</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/11/statistic/statistic3%20-%20Law%20of%20large%20numbers%20and%20central%20limit%20law/" title="statistic0 - StatQuest"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic0 - StatQuest"/></a><div class="content"><a class="title" href="/2023/02/11/statistic/statistic3%20-%20Law%20of%20large%20numbers%20and%20central%20limit%20law/" title="statistic0 - StatQuest">statistic0 - StatQuest</a><time datetime="2023-02-10T16:00:00.000Z" title="Created 2023-02-11 00:00:00">2023-02-11</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2022 - 2023 By pigudog</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div><div class="footer_custom_text">Hi, welcome to my <a target="_blank" rel="noopener" href="https://pigudog.com">blog</a>!</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="Switch Between Traditional Chinese And Simplified Chinese">简</button><button id="darkmode" type="button" title="Switch Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><div class="js-pjax"></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>