<!DOCTYPE html><html lang="en" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>Introduction to Single-cell RNA-seq | PiguDog的小屋</title><meta name="keywords" content="BIA"><meta name="author" content="pigudog,2205638116@qq.com"><meta name="copyright" content="pigudog"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="单细胞转录组实战演练+云服务器处理中单细胞转录组数据需知和答疑 https:&#x2F;&#x2F;docs.qq.com&#x2F;doc&#x2F;DT2NwV0Fab3JBRUx0课程目录和配套习题： https:&#x2F;&#x2F;mp.weixin.qq.com&#x2F;s&#x2F;fE2yPbVvD0R5I8qnNh4F1w github： https:&#x2F;&#x2F;github.com&#x2F;jmzeng1314&#x2F;scRNA_smart_seq2 学习资料  单细胞天地">
<meta property="og:type" content="article">
<meta property="og:title" content="Introduction to Single-cell RNA-seq">
<meta property="og:url" content="http://example.com/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA9/index.html">
<meta property="og:site_name" content="PiguDog的小屋">
<meta property="og:description" content="单细胞转录组实战演练+云服务器处理中单细胞转录组数据需知和答疑 https:&#x2F;&#x2F;docs.qq.com&#x2F;doc&#x2F;DT2NwV0Fab3JBRUx0课程目录和配套习题： https:&#x2F;&#x2F;mp.weixin.qq.com&#x2F;s&#x2F;fE2yPbVvD0R5I8qnNh4F1w github： https:&#x2F;&#x2F;github.com&#x2F;jmzeng1314&#x2F;scRNA_smart_seq2 学习资料  单细胞天地">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png">
<meta property="article:published_time" content="2022-11-16T16:00:00.000Z">
<meta property="article:modified_time" content="2022-11-15T09:17:01.197Z">
<meta property="article:author" content="pigudog">
<meta property="article:tag" content="BIA">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="http://example.com/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA9/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: {"defaultEncoding":1,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: 'days',
  date_suffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: {"limitCount":50,"languages":{"author":"Author: pigudog","link":"Link: ","source":"Source: PiguDog的小屋","info":"Copyright is owned by the author. For commercial reprints, please contact the author for authorization. For non-commercial reprints, please indicate the source."}},
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'Introduction to Single-cell RNA-seq',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2022-11-15 17:17:01'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.2.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/img/avatar.gif" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">69</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">12</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">16</div></a></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> Movie</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">PiguDog的小屋</a></span><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li><li><a class="site-page child" href="/movies/"><i class="fa-fw fas fa-video"></i><span> Movie</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">Introduction to Single-cell RNA-seq<a class="post-edit-link" href="https://github.com/pigudog/_posts/生信分析/BIA9.md" title="Edited on" target="_blank"><i class="fas fa-pencil-alt"></i></a></h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2022-11-16T16:00:00.000Z" title="Created 2022-11-17 00:00:00">2022-11-17</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2022-11-15T09:17:01.197Z" title="Updated 2022-11-15 17:17:01">2022-11-15</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/medical/">medical</a><i class="fas fa-angle-right post-meta-separator"></i><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/medical/BIA/">BIA</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="Introduction to Single-cell RNA-seq"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><p>单细胞转录组实战演练+云服务器处理中单细胞转录组数据<br>需知和答疑 <a target="_blank" rel="noopener" href="https://docs.qq.com/doc/DT2NwV0Fab3JBRUx0">https://docs.qq.com/doc/DT2NwV0Fab3JBRUx0</a><br>课程目录和配套习题： <a target="_blank" rel="noopener" href="https://mp.weixin.qq.com/s/fE2yPbVvD0R5I8qnNh4F1w">https://mp.weixin.qq.com/s/fE2yPbVvD0R5I8qnNh4F1w</a></p>
<p>github： <a target="_blank" rel="noopener" href="https://github.com/jmzeng1314/scRNA_smart_seq2">https://github.com/jmzeng1314/scRNA_smart_seq2</a></p>
<p>学习资料</p>
<ul>
<li>单细胞天地 - 公众号</li>
<li>生物信息学文献阅读笔记 <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/c_1024966446748618752">https://zhuanlan.zhihu.com/c_1024966446748618752</a></li>
<li>生信技能树官网： <a target="_blank" rel="noopener" href="http://www.biotrainee.com/">http://www.biotrainee.com/</a></li>
<li>介绍 <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/535991249">https://zhuanlan.zhihu.com/p/535991249</a></li>
</ul>
<h1 id="上游分析"><a href="#上游分析" class="headerlink" title="上游分析"></a>上游分析</h1><p>参考转录组测序<br><a href="BIA5.md">BIA5</a><br>如果想下载作者测序的原始测序数据：<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/sra?term=SRP133642">SRP133642</a> 来走一波RNA-seq上游流程就需要：<br>shell脚本处理RNA-seq数据上游分析全部代码在：<a target="_blank" rel="noopener" href="https://github.com/jmzeng1314/scRNA_smart_seq2/blob/master/shell.txt">code</a></p>
<p><a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111229">GSE111229</a><br><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA9/Pasted%20image%2020221115161450.png"><br>进入官网后，发现800个细胞，才测了10个G，可见测序的不够好</p>
<ul>
<li>首先是下游分析，可以参考转录组测序的笔记<a href="BIA5.md">BIA5</a></li>
<li>由于只是试验一下，我们就不进行操作了</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br></pre></td><td class="code"><pre><span class="line"># 安装RNA-seq数据处理流程</span><br><span class="line"></span><br><span class="line">wget https://mirrors.tuna.tsinghua.edu.cn/anaconda/miniconda/Miniconda3-latest-Linux-x86_64.sh</span><br><span class="line"></span><br><span class="line">bash Miniconda3-latest-Linux-x86_64.sh</span><br><span class="line"></span><br><span class="line">source ~/.bashrc</span><br><span class="line"></span><br><span class="line">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free</span><br><span class="line"></span><br><span class="line">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge</span><br><span class="line"></span><br><span class="line">conda config --add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda</span><br><span class="line"></span><br><span class="line">conda config --set show_channel_urls yes</span><br><span class="line"></span><br><span class="line">conda create -n rna python=2 #创建名为rna的软件安装环境</span><br><span class="line"></span><br><span class="line">source activate rna</span><br><span class="line"></span><br><span class="line">conda install -y sra-tools multiqc trim-galore subread hisat2</span><br><span class="line"></span><br><span class="line">## 走RNA-seq数据分析流程</span><br><span class="line"></span><br><span class="line">dump=/home/jianmingzeng/biosoft/sratoolkit/sratoolkit.2.9.2-centos_linux64/bin/fastq-dump</span><br><span class="line"></span><br><span class="line"># 如果使用conda安装的 sra-tools ，那么 fastq-dump命令应该是在环境变量的。</span><br><span class="line"></span><br><span class="line">ls ./sra/* | while read id; do ( nohup $dump --gzip --split-3 -O ./ $&#123;id&#125; &amp; ); done ## 批量转换sra到fq格式。</span><br><span class="line"></span><br><span class="line">hisat2=/home/jianmingzeng/biosoft/HISAT/hisat2-2.0.4/hisat2</span><br><span class="line"></span><br><span class="line"># # 如果使用conda安装的 hisat2，那么 hisat2 命令应该是在环境变量的。</span><br><span class="line"></span><br><span class="line">## 索引文件需要自己下载</span><br><span class="line"></span><br><span class="line"># https://ccb.jhu.edu/software/hisat2/manual.shtml</span><br><span class="line"></span><br><span class="line"># wget ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/data/mm10.tar.gz</span><br><span class="line"></span><br><span class="line">index=/home/jianmingzeng/reference/index/hisat/mm10/genome</span><br><span class="line"></span><br><span class="line">ls raw_fq/*gz | while read id; do</span><br><span class="line"></span><br><span class="line">$hisat2 -p 10 -x $index -U $id -S $&#123;id%%.*&#125;.hisat.sam</span><br><span class="line"></span><br><span class="line">done</span><br><span class="line"></span><br><span class="line">ls *.sam|while read id ;do (samtools sort -O bam -@ 5 -o $(basename $&#123;id&#125; &quot;.sam&quot;).bam $&#123;id&#125;);done</span><br><span class="line"></span><br><span class="line">rm *.sam</span><br><span class="line"></span><br><span class="line">ls *.bam |xargs -i samtools index &#123;&#125;</span><br><span class="line"></span><br><span class="line">## gtf文件推荐去gencode数据库下载</span><br><span class="line"></span><br><span class="line">gtf=/home/jianmingzeng/reference/gtf/gencode/gencode.vM12.annotation.gtf</span><br><span class="line"></span><br><span class="line">featureCounts=/home/jianmingzeng/biosoft/featureCounts/subread-1.5.3-Linux-x86_64/bin/featureCounts</span><br><span class="line"></span><br><span class="line"># # # 如果使用conda安装的 subread，那么featureCounts 命令应该是在环境变量的。</span><br><span class="line"></span><br><span class="line">$featureCounts -T 5 -p -t exon -g gene_id -a $gtf -o all.id.txt *.bam 1&gt;counts.id.log 2&gt;&amp;1 &amp;</span><br></pre></td></tr></table></figure>
<ul>
<li>代码参考：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/a84cd44bac67">https://www.jianshu.com/p/a84cd44bac67</a> 已经失效</li>
<li>视频教程见：<a target="_blank" rel="noopener" href="https://www.bilibili.com/video/av28453557">https://www.bilibili.com/video/av28453557</a></li>
</ul>
<h1 id="文献阅读-单细胞转录组探索CAFs的功能和空间异质性"><a href="#文献阅读-单细胞转录组探索CAFs的功能和空间异质性" class="headerlink" title="文献阅读 - 单细胞转录组探索CAFs的功能和空间异质性"></a>文献阅读 - 单细胞转录组探索CAFs的功能和空间异质性</h1><ul>
<li>单细胞的技术的两个方向<ul>
<li>尽可能多的一个细胞的表达量</li>
<li>细胞数量的个数足够多</li>
</ul>
</li>
</ul>
<p>首先我们的例子是2018年12月的NC文章：<a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41467-018-07582-3">Spatially and functionally distinct subclasses of breast cancer-associated fibroblasts revealed by single cell RNA sequencing</a> 使用成熟的单细胞转录组( Smart-seq2 )手段探索了癌相关的成纤维细胞 CAFs的功能和空间异质性。</p>
<p><strong>Cancer-associated fibroblasts</strong> 癌症相关的成纤维细胞 (CAFs) are a major constituent of the tumor microenvironment(是肿瘤微环境的主要组成部分), although their origin and roles in shaping disease initiation, progression and treatment response remain unclear due to significant heterogeneity 异质性 . Here, following a negative selection strategy combined with single-cell RNA sequencing of 768 transcriptomes of mesenchymal cells from a genetically engineered mouse model of breast cancer, we define three distinct subpopulations of CAFs.我们采用负选择策略，结合768个间充质细胞转录组的单细胞RNA测序，从基因工程的乳腺癌小鼠模型，我们定义了三个不同的CAFs亚群。 Validation at the transcriptional and protein level in several experimental models of cancer and human tumors reveal spatial separation of the CAF subclasses attributable to different origins, including the peri-vascular niche, the mammary fat pad and the transformed epithelium. Gene profiles for each CAF subtype correlate to distinctive functional programs and hold independent prognostic capability in clinical cohorts by association to metastatic disease. In conclusion, the improved resolution of the widely defined CAF population opens the possibility for biomarker-driven development of drugs for precision targeting of CAFs.总之，广泛定义的CAF人群的分辨率的提高为生物标志物驱动的精确靶向CAF药物的开发提供了可能。</p>
<h2 id="名词解释"><a href="#名词解释" class="headerlink" title="名词解释"></a>名词解释</h2><ul>
<li><p>single cell RNA sequencing</p>
<ul>
<li>主要解决的问题<ul>
<li>heterogeneity<ul>
<li>鉴定某些因为含量（or proportion）较低而在常规RNA-seq检测时被“掩盖”的细胞亚群；</li>
<li>检测细胞群体中不同的个体细胞：比如表达不同 TCR 的T细胞，胚胎发育早期的各个细胞等；</li>
<li>追踪某群细胞内细胞间的谱系（lineage）或发育关系；</li>
</ul>
</li>
<li>获得基因表达、剪接等信息以及根据这些信息构建的调控关系</li>
</ul>
</li>
<li>主要依赖的技术<ul>
<li>SMART-seq2 技术<ul>
<li>可用于单细胞mRNA测序，对RNA质量要求高，RNA 降解会引起5’端信息丢失，能够获得RNA 全长扩增 cDNA 产物，打断后全部用于文库的构建及测序</li>
</ul>
</li>
<li>10xgenomics 技术<ul>
<li>可用于单细胞 mRNA 测序，对RNA质量要求高，RNA 降解会引起 5’ 端信息丢失，其一次性能够分离500-1000个细胞，并能够将分离的单个细胞中的微量mRNA通过高效扩增后再进行测序，只能测 3’ 端的一小段序列</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><p>Cancer-associated fibroblast(CAFs) &#x2F; tumor supporting CAF -&gt; 当下的热点：肿瘤微环境</p>
<ul>
<li>CAF: 肿瘤相关成纤维细胞是一类具有高度异质性的细胞群，不同的细胞亚群可能起源于不同的前体细胞，如固有成纤维细胞，肿瘤上皮细胞，肿瘤内皮细胞，骨髓来源细胞，其他间充质细胞等</li>
<li>一些列表明CAF通过多种途径在肿瘤发生、进展、转移等过程中其重要作用<ul>
<li>CAF可以积极重塑纤维间质的细胞外基质</li>
<li>细胞外基质蛋白与非肿瘤细胞(如免疫细胞)上的整合素受体结合，将细胞外基质蛋白募集到肿瘤细胞处并发生相互作用，继而诱导肿瘤细胞发生转移侵袭</li>
<li>高密度细胞外基质会增加组织间隙液压，阻碍药物的传递和吸收，致使低浓度化疗药到达肿瘤部位，这是肿瘤失败的原因之一</li>
<li>CAF以细胞因子和生长因子的形式分泌致瘤信号，促进肿瘤细胞生长、增强肿瘤细胞的增殖和迁移，维持肿瘤增殖和迁移，维持肿瘤进展和转移，进来发现，转录调节热休克因子1(HSF1)在CAF中上调，进而生成TGF-beta and SDF-1alpha, 促进肿瘤恶化；</li>
<li>此外，CAF通过分泌HGF和白介素-6(Interleukin-6, IL-6)促进肿瘤对化疗和酪氨酸激酶一直剂产生耐药性</li>
</ul>
</li>
</ul>
</li>
<li><p>extracellular matrix (ECM)</p>
<ul>
<li>细胞外基质是由大分子构成的错综复杂的网络，为细胞的生存和活动提供适宜的场所，并通过信号转导系统影响细胞的形态，代谢，功能，迁移，增殖和分化</li>
<li>构成细胞外基质的大分子种类繁多<ul>
<li>胶原</li>
<li>非胶原糖蛋白</li>
<li>氨基聚糖与蛋白聚糖</li>
<li>以及弹性蛋白</li>
</ul>
</li>
<li>上皮、肌组织和脑和脊髓中的ECM较少，而结缔组织中的ECM含量较高</li>
</ul>
</li>
<li><p>epithelial-to-mesenchymal transition(EMT)</p>
<ul>
<li>胚胎发育和癌症发展中的细胞可塑性变化有惊人的相似性，而这种可塑性变化收到上皮上皮间质转化过程的调节。</li>
<li>胚胎发育时期，上皮状态和间充质状态的细胞能够自由转化。</li>
<li>EMT使得细胞具备转移和浸润特性</li>
<li>其反向过程，间质上皮转化 mesenchymal-epithliatrasition MET 富裕细胞极性变化并是去移动能力</li>
<li>EMT会促进癌细胞从病灶分离，转移到其他部位，而MET导致癌细胞停留，并在停留出引起新的肿瘤</li>
</ul>
</li>
<li><p>MMTV-<strong>PyMT</strong> mouse model</p>
<ul>
<li>自发性肿瘤模型小鼠-乳腺癌MMTV-PyMT转基因小鼠是通过遗传育种而保留下来的一类自发性乳腺癌动物模型</li>
</ul>
</li>
<li><p>a negative selection fluorescence-activated cell sorting(FACS) strategy</p>
<ul>
<li>用于活细胞荧光激活细胞分选术 FACS 根据荧光标记将一个细胞群分为多个亚群，在流式细胞仪中，这种分选的机制相比非分选分析更为复杂</li>
</ul>
</li>
<li><p>smart-seq2 protocol</p>
</li>
<li><p>External RNA Controls Consortium (ERCC) spiked in</p>
<ul>
<li>还没有定论，是一种屁股技术差异的策略</li>
</ul>
</li>
<li><p>principal-component analysis(PCA)</p>
<ul>
<li>线性降维</li>
</ul>
</li>
<li><p>t-distributed stochastic neighbor embedding(t-SNE)</p>
<ul>
<li>更好的将细胞亚群分开是单细胞分析中常用的手段</li>
<li>一种 Manifold Learning method：将数据点之间的相似度转化为概率。</li>
</ul>
</li>
<li><p>reporducibility-optimized test statistic(ROTS)</p>
</li>
<li><p>differentially expressed genes&#x2F;significantly differentially expressed(SDE)</p>
<ul>
<li>从差异分析到表达矩阵：<ul>
<li>如果是 bulk RNA-seq, 那么现在最流行的就是<strong>DESeq2 和 edgeR</strong>啦，而且有很多经过<strong>RT-qPCR检验过的真实测序数据</strong>可以来评价不同的差异基因算法的表现</li>
<li>对于单细胞测序数据来说，首先需要通过<strong>聚类之后将细胞群体进行分组</strong>，然后来比较不同的组的差异表达情况。当然，也有不少单细胞测序实验设计本身就有时间点，不同个体来源，不同培养条件这样的分组。</li>
<li>同时还有不少方式是不需要预先分类的，因为分类本身可能会引入偏差</li>
<li>跟 bulk RNA-seq 不一样的地方是， scRNA-seq通常涉及到样本数量更多。这时候可以使用非参检验算法，比如 Kolmogory-Smirnov test(KS-test)等等</li>
<li>评价指标<ul>
<li>召回率，True positive rate(TPR), TP&#x2F;(TP+FN)</li>
<li>准确率，False positive rate(FRP), FP&#x2F;(FP+TP)</li>
<li>receiver-operating-characterisitic curve ROC</li>
<li>area under this curve AUC</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li><p>log2(RPKM+1)</p>
</li>
<li><p>PAM50 proliferation metagene</p>
</li>
<li><p>METABRIC cohort</p>
</li>
<li><p>The Cancer Genome Atlas(TCGA) database</p>
</li>
<li><p>six commonly used mesenchymal cell markers</p>
</li>
<li><p>endothelial cell metagene in breast tumor</p>
</li>
</ul>
<h1 id="表达矩阵获取"><a href="#表达矩阵获取" class="headerlink" title="表达矩阵获取"></a>表达矩阵获取</h1><p>可以查找到我们的GEO链接：<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111229">GSE111229</a> 就可以找到作者处理好的表达矩阵（counts和RPKM格式的都有）<br><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA9/Pasted%20image%2020221109102351.png"></p>
<ul>
<li>一个大一个小<ul>
<li>counts 是整数，而normalization之后的值不是整数，所以占用的空间也会更大</li>
</ul>
</li>
<li></li>
</ul>
<h2 id="熟悉文献作者提供而两个表达矩阵"><a href="#熟悉文献作者提供而两个表达矩阵" class="headerlink" title="熟悉文献作者提供而两个表达矩阵"></a>熟悉文献作者提供而两个表达矩阵</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  a<span class="operator">=</span>read.table<span class="punctuation">(</span><span class="string">&#x27;../GSE111229_Mammary_Tumor_fibroblasts_768samples_rawCounts.txt.gz&#x27;</span><span class="punctuation">,</span></span><br><span class="line">               header <span class="operator">=</span> <span class="built_in">T</span> <span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&#x27;\t&#x27;</span><span class="punctuation">)</span>  </span><br><span class="line"><span class="comment">##把表达矩阵文件载入R，header=T :保留文件头部信息，seq=&#x27;\t&#x27;以tap为分隔符</span></span><br><span class="line">  <span class="comment"># 每次都要检测数据</span></span><br><span class="line">  a<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">6</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span> <span class="comment">#对于a矩阵取第1~6行，第1~4列</span></span><br><span class="line">  <span class="comment">## 读取RNA-seq的 counts 定量结果，表达矩阵需要进行简单的过滤</span></span><br><span class="line">  dat<span class="operator">=</span>a<span class="punctuation">[</span>apply<span class="punctuation">(</span>a<span class="punctuation">,</span><span class="number">1</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> <span class="built_in">sum</span><span class="punctuation">(</span>x<span class="operator">&gt;</span><span class="number">1</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="built_in">floor</span><span class="punctuation">(</span>ncol<span class="punctuation">(</span>a<span class="punctuation">)</span><span class="operator">/</span><span class="number">50</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span> </span><br><span class="line">  <span class="comment">#筛选表达量合格的行(基因), 列（细胞）数不变</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">##ncol()返回矩阵的列数值；floor()四舍五入取整数；</span></span><br><span class="line">  <span class="comment">##function() 定义一个函数；sum()求和</span></span><br><span class="line">  <span class="comment">#上面的apply()指令代表对矩阵a进行行计算，判断每行表达量&gt;1的样本总个数，并筛选出细胞表达量合格的基因（行）</span></span><br><span class="line">  <span class="comment">#第一个参数是指要参与计算的矩阵——a</span></span><br><span class="line">  <span class="comment">#第二个参数是指按行计算还是按列计算，1——表示按行计算，2——按列计算；</span></span><br><span class="line">  <span class="comment">#第三个参数是指具体的运算参数,定义一个函数x（即表达量为x）</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">#对每行中x&gt;1的列（即样本数）求和，即得出的是每行中表达量大于1的样本数，</span></span><br><span class="line">  <span class="comment">#然后再筛选出大于floor(ncol(a)/50)的行，这样的行（基因）的细胞表达量才算合格</span></span><br><span class="line">  <span class="comment">#因为2%的细胞有表达量，所以对于768个细胞样本，每个行(基因)在细胞中的表达至少要有15.36（约等于15）个样本表达才算合格</span></span><br><span class="line">  <span class="comment">## 2 % 的细胞有表达量</span></span><br><span class="line">  </span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA9/Pasted%20image%2020221115164720.png"></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line"><span class="built_in">sum</span><span class="punctuation">(</span>dat<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">]</span><span class="punctuation">)</span> <span class="comment"># 第三个细胞测了多少reads -&gt; 200000=0.2M</span></span><br><span class="line"><span class="comment">#  0610007P14Rik基因 in  SS2_15_0048_A5细胞 </span></span><br><span class="line">log2<span class="punctuation">(</span><span class="number">18</span><span class="operator">*</span><span class="number">1000000</span><span class="operator">/</span><span class="built_in">sum</span><span class="punctuation">(</span>dat<span class="punctuation">[</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="operator">+</span><span class="number">1</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">## 18 -- &gt; 6.459884  ## SS2_15_0048_A5</span></span><br><span class="line">dat<span class="operator">=</span>log2<span class="punctuation">(</span>edgeR<span class="operator">::</span>cpm<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="operator">+</span><span class="number">1</span><span class="punctuation">)</span> </span><br><span class="line"><span class="comment">##归一化的一种选择，这里是CPM(count-per-million，每百万碱基中每个转录本的count值)</span></span><br><span class="line"><span class="comment">###CPM只对read count相对总reads数做了数量的均一化，去除文库大小差异。</span></span><br><span class="line">dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span> </span><br><span class="line"></span><br><span class="line"><span class="comment">## 下面介绍一下 dist 函数</span></span><br><span class="line">x<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">10</span></span><br><span class="line">y<span class="operator">=</span><span class="number">2</span><span class="operator">*</span>x</span><br><span class="line">z<span class="operator">=</span>rnorm<span class="punctuation">(</span><span class="number">10</span><span class="punctuation">)</span></span><br><span class="line">tmp<span class="operator">=</span>data.frame<span class="punctuation">(</span>x<span class="punctuation">,</span>y<span class="punctuation">,</span>z<span class="punctuation">)</span></span><br><span class="line">dist<span class="punctuation">(</span>tmp<span class="punctuation">)</span></span><br><span class="line">head<span class="punctuation">(</span>tmp<span class="punctuation">)</span></span><br><span class="line">dist<span class="punctuation">(</span>t<span class="punctuation">(</span>tmp<span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment"># 因为dist是按行计算distance的，所以需要转置</span></span><br><span class="line">cor<span class="punctuation">(</span>tmp<span class="punctuation">)</span></span><br><span class="line">dist<span class="punctuation">(</span>t<span class="punctuation">(</span>scale<span class="punctuation">(</span>tmp<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 可以看到dist函数计算样本直接距离和cor函数计算样本直接相关性，是完全不同的概念。虽然我都没有调它们两个函数的默认的参数。</span></span><br><span class="line"><span class="comment"># </span></span><br><span class="line"><span class="comment"># 总结：</span></span><br><span class="line"><span class="comment"># </span></span><br><span class="line"><span class="comment"># - dist函数计算行与行（样本）之间的距离</span></span><br><span class="line"><span class="comment"># - cor函数计算列与列（样本）之间的相关性</span></span><br><span class="line"><span class="comment"># - scale函数默认对每一列（样本）内部归一化</span></span><br><span class="line"><span class="comment"># - 计算dist之前，最好是对每一个样本（列）进行scale一下</span></span><br><span class="line"><span class="comment"># </span></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#层次聚类，因为近 800细胞，非常耗时。</span></span><br><span class="line">hc<span class="operator">=</span>hclust<span class="punctuation">(</span>dist<span class="punctuation">(</span>t<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">##样本间层次聚类</span></span><br><span class="line"><span class="comment"># 原始表达矩阵转置后，细胞在行，所以计算的是细胞与细胞之间的距离。</span></span><br><span class="line"><span class="comment">## statquest 有详细讲解背后的统计学原理。</span></span><br><span class="line"><span class="built_in">class</span><span class="punctuation">(</span>hc<span class="punctuation">)</span></span><br><span class="line"><span class="operator">?</span>plot.hclust</span><br><span class="line"><span class="comment">## 查看说明书。</span></span><br><span class="line">plot<span class="punctuation">(</span>hc<span class="punctuation">,</span>labels <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#t:矩阵转置，行转列，列转行</span></span><br><span class="line"><span class="comment">#分类时常常需要估算不同样本之间的相似性(Similarity Measurement)</span></span><br><span class="line"><span class="comment"># 这时通常采用的方法就是计算样本间”距离”(Distance)。</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#dist函数是R语言计算距离的主要函数。dist函数可以计算行与行两两间的距离。</span></span><br><span class="line"><span class="comment"># 所以之前的矩阵里面行是基因，转置后行是样本，因为我们要计算样本与样本之间的距离。</span></span><br><span class="line"><span class="comment"># dist()函数计算变量间距离</span></span><br><span class="line"><span class="comment">#hclust函数用来层次聚类</span></span><br><span class="line"></span><br><span class="line">clus <span class="operator">=</span> cutree<span class="punctuation">(</span>hc<span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span> <span class="comment">#对hclust()函数的聚类结果进行剪枝，即选择输出指定类别数的系谱聚类结果。</span></span><br><span class="line">group_list<span class="operator">=</span> as.factor<span class="punctuation">(</span>clus<span class="punctuation">)</span> <span class="comment">##转换为因子属性</span></span><br><span class="line">table<span class="punctuation">(</span>group_list<span class="punctuation">)</span> <span class="comment">##统计频数</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#提取批次信息</span></span><br><span class="line">colnames<span class="punctuation">(</span>dat<span class="punctuation">)</span> <span class="comment">#取列名</span></span><br><span class="line">library<span class="punctuation">(</span>stringr<span class="punctuation">)</span></span><br><span class="line">plate<span class="operator">=</span>str_split<span class="punctuation">(</span>colnames<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&#x27;_&#x27;</span><span class="punctuation">,</span>simplify <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">]</span> <span class="comment">#取列名，以&#x27;_&#x27;号分割，提取第三列。</span></span><br><span class="line"><span class="comment">#str_split()函数可以分割字符串</span></span><br><span class="line">table<span class="punctuation">(</span>plate<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<p><img src="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA9/Pasted%20image%2020221115165712.png"></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">  n_g <span class="operator">=</span> apply<span class="punctuation">(</span>a<span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> <span class="built_in">sum</span><span class="punctuation">(</span>x<span class="operator">&gt;</span><span class="number">1</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">#统计每个样本有表达的有多少行（基因）</span></span><br><span class="line">  <span class="comment"># 这里我们定义， reads数量大于1的那些基因为有表达，一般来说单细胞转录组过半数的基因是不会表达的。</span></span><br><span class="line">  <span class="comment"># 而且大部分单细胞转录组技术很烂，通常超过75%的基因都没办法检测到。</span></span><br><span class="line"></span><br><span class="line">  df<span class="operator">=</span>data.frame<span class="punctuation">(</span>g<span class="operator">=</span>group_list<span class="punctuation">,</span>plate<span class="operator">=</span>plate<span class="punctuation">,</span>n_g<span class="operator">=</span>n_g<span class="punctuation">)</span> <span class="comment">#新建数据框(细胞的属性信息)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">##(样本为行名，列分别为：样本分类信息，样本分组，样本表达的基因数【注意：不是表达量的和，而是种类数或者说个数】)</span></span><br><span class="line">  </span><br><span class="line">  df<span class="operator">$</span><span class="built_in">all</span><span class="operator">=</span><span class="string">&#x27;all&#x27;</span> <span class="comment">#添加列，列名为&quot;all&quot;，没事意思，就是后面有需要</span></span><br><span class="line">  metadata<span class="operator">=</span>df</span><br><span class="line">  save<span class="punctuation">(</span>a<span class="punctuation">,</span>dat<span class="punctuation">,</span>df<span class="punctuation">,</span>file <span class="operator">=</span> <span class="string">&#x27;../input.Rdata&#x27;</span><span class="punctuation">)</span> <span class="comment">#保存a,dat,df这变量到上级目录的input.Rdata</span></span><br><span class="line">  <span class="comment"># 因为另外一个项目也需要使用这个数据集，所以保存到了上级目录。</span></span><br><span class="line"><span class="punctuation">&#125;</span> </span><br></pre></td></tr></table></figure>

<p>代码总结</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  a<span class="operator">=</span>read.table<span class="punctuation">(</span><span class="string">&#x27;../GSE111229_Mammary_Tumor_fibroblasts_768samples_rpkmNormalized.txt.gz&#x27;</span><span class="punctuation">,</span></span><br><span class="line">               header <span class="operator">=</span> <span class="built_in">T</span> <span class="punctuation">,</span>sep <span class="operator">=</span> <span class="string">&#x27;\t&#x27;</span><span class="punctuation">)</span>  <span class="comment">##把表达矩阵文件载入R，header=T :保留文件头部信息，seq=&#x27;\t&#x27;以tap为分隔符</span></span><br><span class="line">  <span class="comment"># 每次都要检测数据</span></span><br><span class="line">  a<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">6</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span> <span class="comment">#对于a矩阵取第1~6行，第1~4列</span></span><br><span class="line">  <span class="comment">## 读取RNA-seq的 counts 定量结果，表达矩阵需要进行简单的过滤</span></span><br><span class="line">  dat<span class="operator">=</span>a<span class="punctuation">[</span>apply<span class="punctuation">(</span>a<span class="punctuation">,</span><span class="number">1</span><span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> <span class="built_in">sum</span><span class="punctuation">(</span>x<span class="operator">&gt;</span><span class="number">0</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="built_in">floor</span><span class="punctuation">(</span>ncol<span class="punctuation">(</span>a<span class="punctuation">)</span><span class="operator">/</span><span class="number">50</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span> <span class="comment">#筛选表达量合格的行,列数不变</span></span><br><span class="line"> </span><br><span class="line">  dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span>   </span><br><span class="line">  <span class="comment">#层次聚类</span></span><br><span class="line">  hc<span class="operator">=</span>hclust<span class="punctuation">(</span>dist<span class="punctuation">(</span>t<span class="punctuation">(</span><span class="built_in">log</span><span class="punctuation">(</span>dat<span class="operator">+</span><span class="number">0.1</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">##样本间层次聚类</span></span><br><span class="line">  <span class="comment"># 如果是基因聚类，可以选择 wgcna 等算法 </span></span><br><span class="line">  <span class="comment">## statquest </span></span><br><span class="line">  plot<span class="punctuation">(</span>hc<span class="punctuation">,</span>labels <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line">  clus <span class="operator">=</span> cutree<span class="punctuation">(</span>hc<span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span> <span class="comment">#对hclust()函数的聚类结果进行剪枝，即选择输出指定类别数的系谱聚类结果。</span></span><br><span class="line">  group_list<span class="operator">=</span> as.factor<span class="punctuation">(</span>clus<span class="punctuation">)</span> <span class="comment">##转换为因子属性</span></span><br><span class="line">  table<span class="punctuation">(</span>group_list<span class="punctuation">)</span> <span class="comment">##统计频数</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">#提取批次信息</span></span><br><span class="line">  colnames<span class="punctuation">(</span>dat<span class="punctuation">)</span> <span class="comment">#取列名</span></span><br><span class="line">  library<span class="punctuation">(</span>stringr<span class="punctuation">)</span></span><br><span class="line">  plate<span class="operator">=</span>str_split<span class="punctuation">(</span>colnames<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="punctuation">,</span><span class="string">&#x27;_&#x27;</span><span class="punctuation">,</span>simplify <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span><span class="number">3</span><span class="punctuation">]</span> <span class="comment">#取列名，以&#x27;_&#x27;号分割，提取第三列。</span></span><br><span class="line">  <span class="comment">#str_split()函数可以分割字符串</span></span><br><span class="line">  table<span class="punctuation">(</span>plate<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  n_g <span class="operator">=</span> apply<span class="punctuation">(</span>a<span class="punctuation">,</span><span class="number">2</span><span class="punctuation">,</span><span class="keyword">function</span><span class="punctuation">(</span>x<span class="punctuation">)</span> <span class="built_in">sum</span><span class="punctuation">(</span>x<span class="operator">&gt;</span><span class="number">0</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">#统计每个样本有表达的有多少行（基因）</span></span><br><span class="line">  <span class="comment"># 这里我们定义， reads数量大于1的那些基因为有表达，一般来说单细胞转录组过半数的基因是不会表达的。</span></span><br><span class="line">  <span class="comment"># 而且大部分单细胞转录组技术很烂，通常超过75%的基因都没办法检测到。</span></span><br><span class="line">  </span><br><span class="line">  df<span class="operator">=</span>data.frame<span class="punctuation">(</span>g<span class="operator">=</span>group_list<span class="punctuation">,</span>plate<span class="operator">=</span>plate<span class="punctuation">,</span>n_g<span class="operator">=</span>n_g<span class="punctuation">)</span> <span class="comment">#新建数据框(细胞的属性信息)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">##(样本为行名，列分别为：样本分类信息，样本分组，样本表达的基因数【注意：不是表达量的和，而是种类数或者说个数】)</span></span><br><span class="line">  </span><br><span class="line">  df<span class="operator">$</span><span class="built_in">all</span><span class="operator">=</span><span class="string">&#x27;all&#x27;</span> <span class="comment">#添加列，列名为&quot;all&quot;，没事意思，就是后面有需要</span></span><br><span class="line">  metadata<span class="operator">=</span>df</span><br><span class="line">  save<span class="punctuation">(</span>dat<span class="punctuation">,</span>metadata<span class="punctuation">,</span>file <span class="operator">=</span> <span class="string">&#x27;../input_rpkm.Rdata&#x27;</span><span class="punctuation">)</span> <span class="comment">#保存a,dat,df这变量到上级目录的input.Rdata</span></span><br><span class="line">  <span class="comment"># 因为另外一个项目也需要使用这个数据集，所以保存到了上级目录。</span></span><br><span class="line"><span class="punctuation">&#125;</span> </span><br><span class="line"></span><br><span class="line"></span><br><span class="line">load<span class="punctuation">(</span>file <span class="operator">=</span> <span class="string">&#x27;../input.Rdata&#x27;</span><span class="punctuation">)</span> <span class="comment">##从上级目录载入input.Rdata</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 每次载入以前的变量，都是可以简单检查一下。</span></span><br><span class="line">a<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span> </span><br><span class="line">dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span> </span><br><span class="line">head<span class="punctuation">(</span>df<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">load<span class="punctuation">(</span>file <span class="operator">=</span> <span class="string">&#x27;../input_rpkm.Rdata&#x27;</span><span class="punctuation">)</span> <span class="comment">##从上级目录载入 input_rpkm.Rdata</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 每次载入以前的变量，都是可以简单检查一下。</span></span><br><span class="line"><span class="comment">#a[1:4,1:4] </span></span><br><span class="line">dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span> </span><br><span class="line">head<span class="punctuation">(</span>metadata<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<h2 id="检测单细胞转录组矩阵看内部异质性"><a href="#检测单细胞转录组矩阵看内部异质性" class="headerlink" title="检测单细胞转录组矩阵看内部异质性"></a>检测单细胞转录组矩阵看内部异质性</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">rm<span class="punctuation">(</span><span class="built_in">list</span> <span class="operator">=</span> ls<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">)</span>  <span class="comment">## 魔幻操作，一键清空~</span></span><br><span class="line">options<span class="punctuation">(</span>stringsAsFactors <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">F</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  load<span class="punctuation">(</span>file <span class="operator">=</span> <span class="string">&#x27;../input.Rdata&#x27;</span><span class="punctuation">)</span></span><br><span class="line">  a<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">  head<span class="punctuation">(</span>df<span class="punctuation">)</span> <span class="comment">#head()函数显示操作前面的信息，默认前6行</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">## 载入第0步准备好的表达矩阵，及细胞的一些属性（hclust分群，plate批次，检测到的基因数量）</span></span><br><span class="line">  <span class="comment"># 注意 变量a是原始的counts矩阵，变量 dat是log2CPM后的表达量矩阵。</span></span><br><span class="line">  </span><br><span class="line">  group_list<span class="operator">=</span>df<span class="operator">$</span>g <span class="comment">#&#x27;$&#x27;符，取列，取metadata（df）矩阵的g列,取出层级聚类信息</span></span><br><span class="line">  table<span class="punctuation">(</span>group_list<span class="punctuation">)</span> <span class="comment">##这是全部基因集的聚类分组信息</span></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br></pre></td><td class="code"><pre><span class="line">  <span class="comment">## 要完全掌握本代码，需要理解什么是热图，什么是PCA图，后面会单独讲解。</span></span><br><span class="line">  </span><br><span class="line">  cg<span class="operator">=</span><span class="built_in">names</span><span class="punctuation">(</span>tail<span class="punctuation">(</span>sort<span class="punctuation">(</span>apply<span class="punctuation">(</span>dat<span class="punctuation">,</span><span class="number">1</span><span class="punctuation">,</span>sd<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="number">100</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">##取表达量标准差最大的100行的行名</span></span><br><span class="line">  <span class="comment"># 这个前面演示过一次，  dat=a[apply(a,1, function(x) sum(x&gt;1) &gt; floor(ncol(a)/50)),] #筛选表达量合格的行,列数不变</span></span><br><span class="line">  <span class="comment"># 大家可以对照着理解 apply，需要一定的R语言功底</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">#对dat矩阵每行求标准差，排序，取最后的100行，并获得后100行的行名（探针名）</span></span><br><span class="line">  <span class="comment">#sd()求标准差，对dat矩阵每一行的counts求标准差</span></span><br><span class="line">  <span class="comment">#sort()函数,排序;</span></span><br><span class="line">  <span class="comment">#tail()函数，显示操作对象后面的信息，默认后6行，这里设定取后100行</span></span><br><span class="line">  <span class="comment">#names()函数，获取或设置对象的名称</span></span><br><span class="line">  library<span class="punctuation">(</span>pheatmap<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">##画热图,针对top100的sd的基因集的表达矩阵,没有聚类分组</span></span><br><span class="line">  pheatmap<span class="punctuation">(</span>dat<span class="punctuation">[</span>cg<span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">,</span>show_colnames <span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span>show_rownames <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span></span><br><span class="line">           filename <span class="operator">=</span> <span class="string">&#x27;all_cells_top_100_sd.png&#x27;</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">##针对top100的sd的基因集的表达矩阵，归一化，分组画图</span></span><br><span class="line">  <span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    </span><br><span class="line">    x<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">10</span>;plot<span class="punctuation">(</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    scale<span class="punctuation">(</span>x<span class="punctuation">)</span>;plot<span class="punctuation">(</span>scale<span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment"># scale的目的是使高低的表达更加的明显</span></span><br><span class="line">    </span><br><span class="line">    n<span class="operator">=</span>t<span class="punctuation">(</span>scale<span class="punctuation">(</span>t<span class="punctuation">(</span>dat<span class="punctuation">[</span>cg<span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">#scale()函数去中心化和标准化</span></span><br><span class="line">    <span class="comment">#对每个探针的表达量进行去中心化和标准化</span></span><br><span class="line">    n<span class="punctuation">[</span>n<span class="operator">&gt;</span><span class="number">2</span><span class="punctuation">]</span><span class="operator">=</span><span class="number">2</span> <span class="comment">#矩阵n中归一化后，大于2的项，赋值使之等于2（相当于设置了一个上限）</span></span><br><span class="line">    n<span class="punctuation">[</span>n<span class="operator">&lt;</span> <span class="operator">-</span><span class="number">2</span><span class="punctuation">]</span><span class="operator">=</span> <span class="operator">-</span><span class="number">2</span> <span class="comment">#小于-2的项，赋值使之等于-2（相当于设置了一个下限）</span></span><br><span class="line">    n<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment"># pheatmap(n,show_colnames =F,show_rownames = F)</span></span><br><span class="line">    </span><br><span class="line">    ac<span class="operator">=</span>data.frame<span class="punctuation">(</span>g<span class="operator">=</span>group_list<span class="punctuation">)</span> <span class="comment">#制作细胞（样本）分组矩阵</span></span><br><span class="line">    rownames<span class="punctuation">(</span>ac<span class="punctuation">)</span><span class="operator">=</span>colnames<span class="punctuation">(</span>n<span class="punctuation">)</span> <span class="comment">##ac的行名（样本名）等于n的列名（样本名）</span></span><br><span class="line">    <span class="comment">##判断分组矩阵的行（样本数）和表达矩阵的列（样本数）是否相等</span></span><br><span class="line">    pheatmap<span class="punctuation">(</span>n<span class="punctuation">,</span>show_colnames <span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span>show_rownames <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span></span><br><span class="line">             annotation_col<span class="operator">=</span>ac<span class="punctuation">,</span></span><br><span class="line">             filename <span class="operator">=</span> <span class="string">&#x27;all_cells_top_100_sd_cutree1.png&#x27;</span><span class="punctuation">)</span></span><br><span class="line">    </span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">## 针对top100的sd的基因集的表达矩阵 进行重新 聚类并且分组。</span></span><br><span class="line">  <span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    </span><br><span class="line">    </span><br><span class="line">    n<span class="operator">=</span>t<span class="punctuation">(</span>scale<span class="punctuation">(</span>t<span class="punctuation">(</span>dat<span class="punctuation">[</span>cg<span class="punctuation">,</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    n<span class="punctuation">[</span>n<span class="operator">&gt;</span><span class="number">2</span><span class="punctuation">]</span><span class="operator">=</span><span class="number">2</span></span><br><span class="line">    n<span class="punctuation">[</span>n<span class="operator">&lt;</span> <span class="operator">-</span><span class="number">2</span><span class="punctuation">]</span><span class="operator">=</span> <span class="operator">-</span><span class="number">2</span></span><br><span class="line">    n<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment">##这个聚类分组只是对top100的sd的基因集</span></span><br><span class="line">    hc<span class="operator">=</span>hclust<span class="punctuation">(</span>dist<span class="punctuation">(</span>t<span class="punctuation">(</span>n<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> </span><br><span class="line">    clus <span class="operator">=</span> cutree<span class="punctuation">(</span>hc<span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span></span><br><span class="line">    group_list<span class="operator">=</span>as.factor<span class="punctuation">(</span>clus<span class="punctuation">)</span></span><br><span class="line">    table<span class="punctuation">(</span>group_list<span class="punctuation">)</span> <span class="comment">##这个聚类分组信息是针对top100的sd的基因集的，和针对全部基因集的分组结果不一样</span></span><br><span class="line">    table<span class="punctuation">(</span>group_list<span class="punctuation">,</span>df<span class="operator">$</span>g<span class="punctuation">)</span> <span class="comment">## 其中 metadata$g 是前面步骤针对全部表达矩阵的层次聚类结果。</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment">## 下面针对本次挑选100个基因的表达矩阵的层次聚类结果进行热图展示。</span></span><br><span class="line">    ac<span class="operator">=</span>data.frame<span class="punctuation">(</span>g<span class="operator">=</span>group_list<span class="punctuation">)</span></span><br><span class="line">    rownames<span class="punctuation">(</span>ac<span class="punctuation">)</span><span class="operator">=</span>colnames<span class="punctuation">(</span>n<span class="punctuation">)</span></span><br><span class="line">    pheatmap<span class="punctuation">(</span>n<span class="punctuation">,</span>show_colnames <span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span>show_rownames <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span></span><br><span class="line">             annotation_col<span class="operator">=</span>ac<span class="punctuation">,</span></span><br><span class="line">             filename <span class="operator">=</span> <span class="string">&#x27;all_cells_top_100_sd_cutree_2.png&#x27;</span><span class="punctuation">)</span></span><br><span class="line">    dev.off<span class="punctuation">(</span><span class="punctuation">)</span> <span class="comment">##关闭画板</span></span><br><span class="line">    </span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">###先对整个基因集聚类分组，再对top100的sd的基因集画图</span></span><br><span class="line">  <span class="comment">###和选取top100的sd的基因集，聚类分组再画图，结果聚类分组信息不同</span></span><br><span class="line">  <span class="comment">###两次的聚类分组信息不同，画出的图不同</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">## 下面是画PCA的必须操作，需要看不同做PCA的包的说明书。</span></span><br><span class="line">  <span class="comment">## 如果想了解PCA分析原理，需要阅读：https://mp.weixin.qq.com/s/Kw05PWD2m65TZu2Blhnl4w</span></span><br><span class="line">  </span><br><span class="line">  dat_back<span class="operator">=</span>dat <span class="comment">##防止下面操作把数值搞坏的一个备份</span></span><br><span class="line">  dat<span class="operator">=</span>dat_back  <span class="comment">##表达矩阵数据</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">  dat<span class="operator">=</span>t<span class="punctuation">(</span>dat<span class="punctuation">)</span></span><br><span class="line">  dat<span class="operator">=</span>as.data.frame<span class="punctuation">(</span>dat<span class="punctuation">)</span> <span class="comment">##转换为数据框</span></span><br><span class="line">  dat<span class="operator">=</span>cbind<span class="punctuation">(</span>dat<span class="punctuation">,</span>group_list <span class="punctuation">)</span> <span class="comment">##cbind()合并列（横向追加）;添加分组信息</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">  <span class="comment">## 表达矩阵可以随心所欲的取行列，基础知识需要打牢。</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">12197</span><span class="operator">:</span><span class="number">12199</span><span class="punctuation">]</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="punctuation">,</span>ncol<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="punctuation">]</span> <span class="comment">#ncol()列，返回列长值</span></span><br><span class="line">  table<span class="punctuation">(</span>dat<span class="operator">$</span>group_list<span class="punctuation">)</span></span><br><span class="line">  library<span class="punctuation">(</span><span class="string">&quot;FactoMineR&quot;</span><span class="punctuation">)</span></span><br><span class="line">  library<span class="punctuation">(</span><span class="string">&quot;factoextra&quot;</span><span class="punctuation">)</span> </span><br><span class="line">  <span class="comment"># The variable group_list (index = ) is removed</span></span><br><span class="line">  <span class="comment"># before PCA analysis</span></span><br><span class="line">  <span class="comment">## 这里的PCA分析，被该R包包装成一个简单的函数，复杂的原理后面讲解。</span></span><br><span class="line">  dat.pca <span class="operator">&lt;-</span> PCA<span class="punctuation">(</span>dat<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span>ncol<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">,</span> graph <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span> <span class="comment">#&#x27;-&#x27;表示“非”</span></span><br><span class="line">  fviz_pca_ind<span class="punctuation">(</span>dat.pca<span class="punctuation">,</span>repel <span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span></span><br><span class="line">               geom.ind <span class="operator">=</span> <span class="string">&quot;point&quot;</span><span class="punctuation">,</span> <span class="comment"># show points only (nbut not &quot;text&quot;)只显示点不显示文本</span></span><br><span class="line">               col.ind <span class="operator">=</span> dat<span class="operator">$</span>group_list<span class="punctuation">,</span> <span class="comment"># color by groups 颜色组</span></span><br><span class="line">               <span class="comment"># palette = c(&quot;#00AFBB&quot;, &quot;#E7B800&quot;),</span></span><br><span class="line">               addEllipses <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> <span class="comment"># Concentration ellipses 集中成椭圆</span></span><br><span class="line">               legend.title <span class="operator">=</span> <span class="string">&quot;Groups&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line">  <span class="comment">## 事实上还是有很多基因dropout非常严重。</span></span><br><span class="line">  ggsave<span class="punctuation">(</span><span class="string">&#x27;all_cells_PCA.png&#x27;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">rm<span class="punctuation">(</span><span class="built_in">list</span> <span class="operator">=</span> ls<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">)</span>  <span class="comment">## 魔幻操作，一键清空~</span></span><br><span class="line">options<span class="punctuation">(</span>stringsAsFactors <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">F</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  </span><br><span class="line">  load<span class="punctuation">(</span>file <span class="operator">=</span> <span class="string">&#x27;../input_rpkm.Rdata&#x27;</span><span class="punctuation">)</span></span><br><span class="line">  <span class="comment"># a[1:4,1:4]</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">  head<span class="punctuation">(</span>metadata<span class="punctuation">)</span> <span class="comment">#head()函数显示操作前面的信息，默认前6行</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">## 载入第0步准备好的表达矩阵，及细胞的一些属性（hclust分群，plate批次，检测到的基因数量）</span></span><br><span class="line">  <span class="comment"># 注意 变量a是原始的counts矩阵，变量 dat是log2CPM后的表达量矩阵。</span></span><br><span class="line">  </span><br><span class="line">  group_list<span class="operator">=</span>metadata<span class="operator">$</span>g <span class="comment">#&#x27;$&#x27;符，取列，取metadata矩阵的g列,取出层级聚类信息</span></span><br><span class="line">  table<span class="punctuation">(</span>group_list<span class="punctuation">)</span> <span class="comment">##这是全部基因集的聚类分组信息</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">## 要完全掌握本代码，需要理解什么是热图，什么是PCA图，后面会单独讲解。</span></span><br><span class="line">  </span><br><span class="line">  cg<span class="operator">=</span><span class="built_in">names</span><span class="punctuation">(</span>tail<span class="punctuation">(</span>sort<span class="punctuation">(</span>apply<span class="punctuation">(</span>dat<span class="punctuation">,</span><span class="number">1</span><span class="punctuation">,</span>sd<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="number">100</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">##取表达量标准差最大的100行的行名</span></span><br><span class="line">  <span class="comment"># 这个前面演示过一次，  dat=a[apply(a,1, function(x) sum(x&gt;1) &gt; floor(ncol(a)/50)),] #筛选表达量合格的行,列数不变</span></span><br><span class="line">  <span class="comment"># 大家可以对照着理解 apply，需要一定的R语言功底</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">#对dat矩阵每行求标准差，排序，取最后的100行，并获得后100行的行名（探针名）</span></span><br><span class="line">  <span class="comment">#sd()求标准差，对dat矩阵每一行的counts求标准差</span></span><br><span class="line">  <span class="comment">#sort()函数,排序;</span></span><br><span class="line">  <span class="comment">#tail()函数，显示操作对象后面的信息，默认后6行，这里设定取后100行</span></span><br><span class="line">  <span class="comment">#names()函数，获取或设置对象的名称</span></span><br><span class="line">  library<span class="punctuation">(</span>pheatmap<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  mat<span class="operator">=</span>log2<span class="punctuation">(</span>dat<span class="punctuation">[</span>cg<span class="punctuation">,</span><span class="punctuation">]</span><span class="operator">+</span><span class="number">0.01</span><span class="punctuation">)</span></span><br><span class="line">  <span class="comment">##画热图,针对top100的sd的基因集的表达矩阵,没有聚类分组</span></span><br><span class="line">  pheatmap<span class="punctuation">(</span>mat<span class="punctuation">,</span>show_colnames <span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span>show_rownames <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span></span><br><span class="line">           filename <span class="operator">=</span> <span class="string">&#x27;all_cells_top_100_sd.png&#x27;</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">##针对top100的sd的基因集的表达矩阵，归一化，分组画图</span></span><br><span class="line">  <span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    </span><br><span class="line">    x<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">10</span>;plot<span class="punctuation">(</span><span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    scale<span class="punctuation">(</span>x<span class="punctuation">)</span>;plot<span class="punctuation">(</span>scale<span class="punctuation">(</span>x<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    </span><br><span class="line">    n<span class="operator">=</span>t<span class="punctuation">(</span>scale<span class="punctuation">(</span>t<span class="punctuation">(</span> mat <span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">#scale()函数去中心化和标准化</span></span><br><span class="line">    <span class="comment">#对每个探针的表达量进行去中心化和标准化</span></span><br><span class="line">    n<span class="punctuation">[</span>n<span class="operator">&gt;</span><span class="number">2</span><span class="punctuation">]</span><span class="operator">=</span><span class="number">2</span> <span class="comment">#矩阵n中归一化后，大于2的项，赋值使之等于2（相当于设置了一个上限）</span></span><br><span class="line">    n<span class="punctuation">[</span>n<span class="operator">&lt;</span> <span class="operator">-</span><span class="number">2</span><span class="punctuation">]</span><span class="operator">=</span> <span class="operator">-</span><span class="number">2</span> <span class="comment">#小于-2的项，赋值使之等于-2（相当于设置了一个下限）</span></span><br><span class="line">    n<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment"># pheatmap(n,show_colnames =F,show_rownames = F)</span></span><br><span class="line">    </span><br><span class="line">    ac<span class="operator">=</span>data.frame<span class="punctuation">(</span>g<span class="operator">=</span>group_list<span class="punctuation">)</span> <span class="comment">#制作细胞（样本）分组矩阵</span></span><br><span class="line">    rownames<span class="punctuation">(</span>ac<span class="punctuation">)</span><span class="operator">=</span>colnames<span class="punctuation">(</span>n<span class="punctuation">)</span> <span class="comment">##ac的行名（样本名）等于n的列名（样本名）</span></span><br><span class="line">    <span class="comment">##判断分组矩阵的行（样本数）和表达矩阵的列（样本数）是否相等</span></span><br><span class="line">    pheatmap<span class="punctuation">(</span>n<span class="punctuation">,</span>show_colnames <span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span>show_rownames <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span></span><br><span class="line">             annotation_col<span class="operator">=</span>ac<span class="punctuation">,</span></span><br><span class="line">             filename <span class="operator">=</span> <span class="string">&#x27;all_cells_top_100_sd_cutree1.png&#x27;</span><span class="punctuation">)</span></span><br><span class="line">    </span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">## 针对top100的sd的基因集的表达矩阵 进行重新 聚类并且分组。</span></span><br><span class="line">  <span class="keyword">if</span><span class="punctuation">(</span><span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">    </span><br><span class="line">    </span><br><span class="line">    n<span class="operator">=</span>t<span class="punctuation">(</span>scale<span class="punctuation">(</span>t<span class="punctuation">(</span> mat <span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    n<span class="punctuation">[</span>n<span class="operator">&gt;</span><span class="number">2</span><span class="punctuation">]</span><span class="operator">=</span><span class="number">2</span></span><br><span class="line">    n<span class="punctuation">[</span>n<span class="operator">&lt;</span> <span class="operator">-</span><span class="number">2</span><span class="punctuation">]</span><span class="operator">=</span> <span class="operator">-</span><span class="number">2</span></span><br><span class="line">    n<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment">##这个聚类分组只是对top100的sd的基因集</span></span><br><span class="line">    hc<span class="operator">=</span>hclust<span class="punctuation">(</span>dist<span class="punctuation">(</span>t<span class="punctuation">(</span>n<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">)</span> </span><br><span class="line">    clus <span class="operator">=</span> cutree<span class="punctuation">(</span>hc<span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span></span><br><span class="line">    group_list<span class="operator">=</span>as.factor<span class="punctuation">(</span>clus<span class="punctuation">)</span></span><br><span class="line">    table<span class="punctuation">(</span>group_list<span class="punctuation">)</span> <span class="comment">##这个聚类分组信息是针对top100的sd的基因集的，和针对全部基因集的分组结果不一样</span></span><br><span class="line">    table<span class="punctuation">(</span>group_list<span class="punctuation">,</span>metadata<span class="operator">$</span>g<span class="punctuation">)</span> <span class="comment">## 其中 metadata$g 是前面步骤针对全部表达矩阵的层次聚类结果。</span></span><br><span class="line">    </span><br><span class="line">    <span class="comment">## 下面针对本次挑选100个基因的表达矩阵的层次聚类结果进行热图展示。</span></span><br><span class="line">    ac<span class="operator">=</span>data.frame<span class="punctuation">(</span>g<span class="operator">=</span>group_list<span class="punctuation">)</span></span><br><span class="line">    rownames<span class="punctuation">(</span>ac<span class="punctuation">)</span><span class="operator">=</span>colnames<span class="punctuation">(</span>n<span class="punctuation">)</span></span><br><span class="line">    pheatmap<span class="punctuation">(</span>n<span class="punctuation">,</span>show_colnames <span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span>show_rownames <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span></span><br><span class="line">             annotation_col<span class="operator">=</span>ac<span class="punctuation">,</span></span><br><span class="line">             filename <span class="operator">=</span> <span class="string">&#x27;all_cells_top_100_sd_cutree_2.png&#x27;</span><span class="punctuation">)</span></span><br><span class="line">    dev.off<span class="punctuation">(</span><span class="punctuation">)</span> <span class="comment">##关闭画板</span></span><br><span class="line">    </span><br><span class="line">  <span class="punctuation">&#125;</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">###先对整个基因集聚类分组，再对top100的sd的基因集画图</span></span><br><span class="line">  <span class="comment">###和选取top100的sd的基因集，聚类分组再画图，结果聚类分组信息不同</span></span><br><span class="line">  <span class="comment">###两次的聚类分组信息不同，画出的图不同</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">## 下面是画PCA的必须操作，需要看不同做PCA的包的说明书。</span></span><br><span class="line">  <span class="comment">## 如果想了解PCA分析原理，需要阅读：https://mp.weixin.qq.com/s/Kw05PWD2m65TZu2Blhnl4w</span></span><br><span class="line">  </span><br><span class="line">  dat_back<span class="operator">=</span>dat <span class="comment">##防止下面操作把数值搞坏的一个备份</span></span><br><span class="line">  dat<span class="operator">=</span>dat_back  <span class="comment">##表达矩阵数据</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">  dat<span class="operator">=</span>t<span class="punctuation">(</span>dat<span class="punctuation">)</span></span><br><span class="line">  dat<span class="operator">=</span>as.data.frame<span class="punctuation">(</span>dat<span class="punctuation">)</span> <span class="comment">##转换为数据框</span></span><br><span class="line">  dat<span class="operator">=</span>cbind<span class="punctuation">(</span>dat<span class="punctuation">,</span>group_list <span class="punctuation">)</span> <span class="comment">##cbind()合并列（横向追加）;添加分组信息</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">]</span></span><br><span class="line">  <span class="comment">## 表达矩阵可以随心所欲的取行列，基础知识需要打牢。</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">4</span><span class="punctuation">,</span><span class="number">12197</span><span class="operator">:</span><span class="number">12199</span><span class="punctuation">]</span></span><br><span class="line">  dat<span class="punctuation">[</span><span class="punctuation">,</span>ncol<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="punctuation">]</span> <span class="comment">#ncol()列，返回列长值</span></span><br><span class="line">  table<span class="punctuation">(</span>dat<span class="operator">$</span>group_list<span class="punctuation">)</span></span><br><span class="line">  library<span class="punctuation">(</span><span class="string">&quot;FactoMineR&quot;</span><span class="punctuation">)</span></span><br><span class="line">  library<span class="punctuation">(</span><span class="string">&quot;factoextra&quot;</span><span class="punctuation">)</span> </span><br><span class="line">  <span class="comment"># The variable group_list (index = ) is removed</span></span><br><span class="line">  <span class="comment"># before PCA analysis</span></span><br><span class="line">  <span class="comment">## 这里的PCA分析，被该R包包装成一个简单的函数，复杂的原理后面讲解。</span></span><br><span class="line">  dat.pca <span class="operator">&lt;-</span> PCA<span class="punctuation">(</span>dat<span class="punctuation">[</span><span class="punctuation">,</span><span class="operator">-</span>ncol<span class="punctuation">(</span>dat<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">,</span> graph <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span> <span class="comment">#&#x27;-&#x27;表示“非”</span></span><br><span class="line">  fviz_pca_ind<span class="punctuation">(</span>dat.pca<span class="punctuation">,</span>repel <span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span></span><br><span class="line">               geom.ind <span class="operator">=</span> <span class="string">&quot;point&quot;</span><span class="punctuation">,</span> <span class="comment"># show points only (nbut not &quot;text&quot;)只显示点不显示文本</span></span><br><span class="line">               col.ind <span class="operator">=</span> dat<span class="operator">$</span>group_list<span class="punctuation">,</span> <span class="comment"># color by groups 颜色组</span></span><br><span class="line">               <span class="comment"># palette = c(&quot;#00AFBB&quot;, &quot;#E7B800&quot;),</span></span><br><span class="line">               addEllipses <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> <span class="comment"># Concentration ellipses 集中成椭圆</span></span><br><span class="line">               legend.title <span class="operator">=</span> <span class="string">&quot;Groups&quot;</span></span><br><span class="line">  <span class="punctuation">)</span></span><br><span class="line">  <span class="comment">## 事实上还是有很多基因dropout非常严重。</span></span><br><span class="line">  ggsave<span class="punctuation">(</span><span class="string">&#x27;all_cells_PCA.png&#x27;</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  </span><br><span class="line">  </span><br><span class="line">  </span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"></span><br></pre></td></tr></table></figure>


<h1 id="转录组分析回顾"><a href="#转录组分析回顾" class="headerlink" title="转录组分析回顾"></a>转录组分析回顾</h1><p>这一单元代码都在 <code>section01-RNA-seq</code> 文件夹，进入打开后缀是 <code>Rproj</code> 的文件就会自动调用你系统的Rstudio软件，从而定位到项目。<br>目录</p>
<ul>
<li>step0-index.R<ul>
<li>读取 作者的counts文件，简单过滤，并且logCPM转换</li>
</ul>
</li>
<li>step1-check.R<ul>
<li>检测表达矩阵里面细胞的相关性，hclust结果，热图展现，PCA图展现</li>
</ul>
</li>
<li>step2-cv2.R<ul>
<li>检测基因的变异系数及基因表达量的均值的相关性，探索其它统计学指标</li>
</ul>
</li>
<li>step3-batch-PCA-tSNE.R<ul>
<li>查看表达量是否受批次效应影响，这里简短介绍了PCA和tSNE作用</li>
</ul>
</li>
<li>step4-gene-number.R<ul>
<li>查看不同分组条件下检测到的基因数量的多少分布</li>
</ul>
</li>
<li>step5-pam50.R<ul>
<li>生物学背景知识，对乳腺癌研究不感兴趣的请不要学习</li>
</ul>
</li>
<li>step6-cell-cycle.R<ul>
<li>生物学背景知识，最好是自行搜索了解</li>
</ul>
</li>
<li>step7-counts2rpkm.R<ul>
<li>完全的R代码技巧，能力不够着请不要浪费时间学习，非常辛苦的。</li>
</ul>
</li>
<li>step8-DEG.R<ul>
<li>根据文章</li>
</ul>
</li>
</ul>
<h2 id="环境配置"><a href="#环境配置" class="headerlink" title="环境配置"></a>环境配置</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">rm<span class="punctuation">(</span><span class="built_in">list</span> <span class="operator">=</span> ls<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">)</span>  <span class="comment">## 魔幻操作，一键清空~</span></span><br><span class="line">Sys.setenv<span class="punctuation">(</span>R_MAX_NUM_DLLS<span class="operator">=</span><span class="number">999</span><span class="punctuation">)</span> <span class="comment">##Sys.setenv修改环境设置，R的namespace是有上限的，如果导入包时超过这个上次就会报错,R_MAX_NUM_DLLS可以修改这个上限</span></span><br><span class="line">options<span class="punctuation">(</span>stringsAsFactors <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span> <span class="comment">##options:允许用户对工作空间进行全局设置，stringsAsFactors防止R自动把字符串string的列辨认成factor</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># http://www.bio-info-trainee.com/3727.html 周末生物信息学培训班准备工作</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">options<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">$</span>repos  <span class="comment">## 查看使用install.packages安装时的默认镜像</span></span><br><span class="line">options<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">$</span>BioC_mirror <span class="comment">##查看使用bioconductor的默认镜像</span></span><br><span class="line">options<span class="punctuation">(</span>BioC_mirror<span class="operator">=</span><span class="string">&quot;https://mirrors.ustc.edu.cn/bioc/&quot;</span><span class="punctuation">)</span> <span class="comment">##指定镜像，这个是中国科技大学镜像</span></span><br><span class="line">options<span class="punctuation">(</span><span class="string">&quot;repos&quot;</span> <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span>CRAN<span class="operator">=</span><span class="string">&quot;https://mirrors.tuna.tsinghua.edu.cn/CRAN/&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="comment">##指定install.packages安装镜像，这个是清华镜像</span></span><br><span class="line">options<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">$</span>repos </span><br><span class="line">options<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">$</span>BioC_mirror</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">if(F)&#123;</span><br><span class="line">  </span><br><span class="line">  </span><br><span class="line">  BiocManager::install(c( &#x27;scran&#x27;),ask = F,update = F)</span><br><span class="line">  </span><br><span class="line">  BiocManager::install(&quot;TxDb.Mmusculus.UCSC.mm10.knownGene&quot;,ask = F,update = F)</span><br><span class="line">  BiocManager::install(&quot;org.Mm.eg.db&quot;,ask = F,update = F)</span><br><span class="line">  BiocManager::install(&quot;genefu&quot;,ask = F,update = F)</span><br><span class="line">  BiocManager::install(&quot;org.Hs.eg.db&quot;,ask = F,update = F)</span><br><span class="line">  BiocManager::install(&quot;TxDb.Hsapiens.UCSC.hg38.knownGene&quot;,ask = F,update = F)</span><br><span class="line">  install.packages(&quot;ggfortify&quot;)</span><br><span class="line">  install.packages(&quot;FactoMineR&quot;)</span><br><span class="line">  install.packages(&quot;factoextra&quot;)</span><br><span class="line">  install.packages(&quot;ggfortify&quot;)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h1 id="单细胞转录组数据分析"><a href="#单细胞转录组数据分析" class="headerlink" title="单细胞转录组数据分析"></a>单细胞转录组数据分析</h1><h2 id="目录"><a href="#目录" class="headerlink" title="目录"></a>目录</h2><h4 id="首先学习3个R包"><a href="#首先学习3个R包" class="headerlink" title="首先学习3个R包"></a>首先学习3个R包</h4><p>见 section02-scRNA 文件夹代码，因为文章其实并没有使用主流单细胞转录组R包，这里仅仅是根据scRNAseq的示例数据来讲解，进入打开后缀是 <code>Rproj</code> 的文件就会自动调用你系统的Rstudio软件，从而定位到项目。</p>
<h4 id="然后尝试把这3个R包应用到该文章的数据"><a href="#然后尝试把这3个R包应用到该文章的数据" class="headerlink" title="然后尝试把这3个R包应用到该文章的数据"></a>然后尝试把这3个R包应用到该文章的数据</h4><p>见 section03-for_paper 文件夹代码，进入打开后缀是 <code>Rproj</code> 的文件就会自动调用你系统的Rstudio软件，从而定位到项目。可以完全复现文章图表。</p>
<h4 id="最后是公共数据库挖掘"><a href="#最后是公共数据库挖掘" class="headerlink" title="最后是公共数据库挖掘"></a>最后是公共数据库挖掘</h4><p>在 section04-downstream 文件夹，进入打开后缀是 <code>Rproj</code> 的文件就会自动调用你系统的Rstudio软件，从而定位到项目。</p>
<h2 id><a href="#" class="headerlink" title></a></h2><h1 id="学习笔记在单细胞天地持续更新"><a href="#学习笔记在单细胞天地持续更新" class="headerlink" title="学习笔记在单细胞天地持续更新"></a>学习笔记在单细胞天地持续更新</h1><ul>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484464&idx=1&sn=cec394967fcd20a25c258bed892f2457&chksm=ea1f02b2dd688ba4c2bb0659f7f66349c49d876ec340c84763c29c7bf080dba5cfea90d4b68e&scene=21#wechat_redirect">单细胞转录组学习笔记-1</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484484&idx=1&sn=eb28c69645f8dbfda6c8a4afaf931883&chksm=ea1f02c6dd688bd0ca7ab1caa3261d9c0f59dc9c83002ca3780bf88e11a94f0893bff3849900&scene=21#wechat_redirect">单细胞转录组学习笔记-2</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484498&idx=1&sn=45a7aa58d726b68dac1c80d4b686f386&chksm=ea1f02d0dd688bc64fd14e5f0370ec1889215fc5ef3837e34db84f1319f517598d3e69832fc2&scene=21#wechat_redirect">单细胞转录组上游分析之shell回顾</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484551&idx=1&sn=f58786c646cad53047b28cdc4f9bcd9e&chksm=ea1f0205dd688b1326aa5309e506452aacaa9edadf46e44d8cc9547f1126abc553362ed96800&scene=21#wechat_redirect">获取Github代码包以及准备工作</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484603&idx=1&sn=6c13a90be81285a71b915c77caa30407&chksm=ea1f0239dd688b2fce040852e8e799d410517e0924c45bb4cd00f749cafbe3972d8def3c2ead&scene=21#wechat_redirect">常说的表达矩阵，那得到之后呢？</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484644&idx=1&sn=eb86415a435da0ef81c1c0ae3c2767fc&chksm=ea1f0266dd688b7030db7bb57782490f0f23d91c79bb9b89a3e6135657544dd131a9cc26a503&scene=21#wechat_redirect">由表达矩阵看内部异质性</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484658&idx=1&sn=c0c42921ad194abb76ebdd1bc800a7f7&chksm=ea1f0270dd688b66eb730b2459eb65b885fe85a2f7478c26f2aafc3b3ae6a5054f4de75d7c2c&scene=21#wechat_redirect">重复平均表达量和变异系数相关性散点图</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484689&idx=1&sn=6689a065d287b0308354b00c0a93bb13&chksm=ea1f0393dd688a85faa81c61c8c2528290a00aa8e379da1f939f5d4f5cfa2127581228961835&scene=21#wechat_redirect">聚类算法之PCA与tSNE</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484705&idx=1&sn=69b50b56286cd09bc1e9462b3218722d&chksm=ea1f03a3dd688ab57971f11b5e6433e0951cc68fff4de007d01f2ee3237413832420b7ad771b&scene=21#wechat_redirect">统计细胞检测的基因数量</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484714&idx=1&sn=c5a232af5b0be717b40af56f9f10f9e2&chksm=ea1f03a8dd688abe0c61699db9d9fb1248c5e749abc9354677662a8fa216f3c78ac029f0fd90&scene=21#wechat_redirect">乳腺癌领域之PAM50分类</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484719&idx=1&sn=c2e28e616261078936e7d34505b03120&chksm=ea1f03addd688abb0e67f1416aef16167584de2b2231b242e1c612e6b873b02eccf1f9575d92&scene=21#wechat_redirect">生物学背景知识之细胞周期推断</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484749&idx=1&sn=f734117fc9fcb604071b9b6df1739aeb&chksm=ea1f03cfdd688ad97a021f34d0b54fba7c253b7ea7c49cd00a6dc59ead3f2359b794b29576c7&scene=21#wechat_redirect">RPKM概念及计算方法</a></li>
<li><a target="_blank" rel="noopener" href="http://mp.weixin.qq.com/s?__biz=MzI1Njk4ODE0MQ==&mid=2247484838&idx=1&sn=2823e03598d88fd4d219988e8bd69397&chksm=ea1f0324dd688a328ad4fe11320815121461ca6b9452cf8e98aa069280931515f8638a776694&scene=21#wechat_redirect">差异分析及KEGG注释简介</a></li>
<li>更新中</li>
</ul>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="http://example.com">pigudog</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://example.com/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA9/">http://example.com/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA9/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/BIA/">BIA</a></div><div class="post_share"><div class="social-share" data-image="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><div class="post-reward"><div class="reward-button"><i class="fas fa-qrcode"></i> Donate</div><div class="reward-main"><ul class="reward-all"><li class="reward-item"><a href="/img/wechat.jpg" target="_blank"><img class="post-qr-code-img" src="/img/wechat.jpg" alt="wechat"/></a><div class="post-qr-code-desc">wechat</div></li></ul></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA6/"><img class="prev-cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">Introduction to WES</div></div></a></div><div class="next-post pull-right"><a href="/2022/11/17/%E7%94%9F%E4%BF%A1%E5%AD%A6%E5%BE%92/BItest1/"><img class="next-cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">BItest1 - 学会解析文章架构</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2022/11/15/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA10/" title="Introduction to ATAC-seq"><img class="cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2022-11-15</div><div class="title">Introduction to ATAC-seq</div></div></a></div><div><a href="/2022/08/01/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA1/" title="BIA1-利用Python进行生信分析"><img class="cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2022-08-01</div><div class="title">BIA1-利用Python进行生信分析</div></div></a></div><div><a href="/2022/10/25/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA2/" title="first view for RNA-seq"><img class="cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2022-10-25</div><div class="title">first view for RNA-seq</div></div></a></div><div><a href="/2022/11/01/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA3/" title="Cloud Server for BIA"><img class="cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2022-11-01</div><div class="title">Cloud Server for BIA</div></div></a></div><div><a href="/2022/11/08/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA4/" title="Intro to GEO"><img class="cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2022-11-08</div><div class="title">Intro to GEO</div></div></a></div><div><a href="/2022/11/08/%E7%94%9F%E4%BF%A1%E5%88%86%E6%9E%90/BIA5/" title="Intro to RNA-seq"><img class="cover" src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2022-11-08</div><div class="title">Intro to RNA-seq</div></div></a></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/img/avatar.gif" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">pigudog</div><div class="author-info__description">冲冲冲！</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">69</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">12</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">16</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/pigudog"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="https://github.com/pigudog" target="_blank" title="Github"><i class="fab fa-github"></i></a><a class="social-icon" href="/2205638116@qq.com" target="_blank" title="Email"><i class="fas fa-envelope"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E4%B8%8A%E6%B8%B8%E5%88%86%E6%9E%90"><span class="toc-number">1.</span> <span class="toc-text">上游分析</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E6%96%87%E7%8C%AE%E9%98%85%E8%AF%BB-%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%8E%A2%E7%B4%A2CAFs%E7%9A%84%E5%8A%9F%E8%83%BD%E5%92%8C%E7%A9%BA%E9%97%B4%E5%BC%82%E8%B4%A8%E6%80%A7"><span class="toc-number">2.</span> <span class="toc-text">文献阅读 - 单细胞转录组探索CAFs的功能和空间异质性</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%90%8D%E8%AF%8D%E8%A7%A3%E9%87%8A"><span class="toc-number">2.1.</span> <span class="toc-text">名词解释</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5%E8%8E%B7%E5%8F%96"><span class="toc-number">3.</span> <span class="toc-text">表达矩阵获取</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%86%9F%E6%82%89%E6%96%87%E7%8C%AE%E4%BD%9C%E8%80%85%E6%8F%90%E4%BE%9B%E8%80%8C%E4%B8%A4%E4%B8%AA%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5"><span class="toc-number">3.1.</span> <span class="toc-text">熟悉文献作者提供而两个表达矩阵</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%A3%80%E6%B5%8B%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E7%9F%A9%E9%98%B5%E7%9C%8B%E5%86%85%E9%83%A8%E5%BC%82%E8%B4%A8%E6%80%A7"><span class="toc-number">3.2.</span> <span class="toc-text">检测单细胞转录组矩阵看内部异质性</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%E5%9B%9E%E9%A1%BE"><span class="toc-number">4.</span> <span class="toc-text">转录组分析回顾</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%8E%AF%E5%A2%83%E9%85%8D%E7%BD%AE"><span class="toc-number">4.1.</span> <span class="toc-text">环境配置</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90"><span class="toc-number">5.</span> <span class="toc-text">单细胞转录组数据分析</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%9B%AE%E5%BD%95"><span class="toc-number">5.1.</span> <span class="toc-text">目录</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#%E9%A6%96%E5%85%88%E5%AD%A6%E4%B9%A03%E4%B8%AAR%E5%8C%85"><span class="toc-number">5.1.0.1.</span> <span class="toc-text">首先学习3个R包</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E7%84%B6%E5%90%8E%E5%B0%9D%E8%AF%95%E6%8A%8A%E8%BF%993%E4%B8%AAR%E5%8C%85%E5%BA%94%E7%94%A8%E5%88%B0%E8%AF%A5%E6%96%87%E7%AB%A0%E7%9A%84%E6%95%B0%E6%8D%AE"><span class="toc-number">5.1.0.2.</span> <span class="toc-text">然后尝试把这3个R包应用到该文章的数据</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E6%9C%80%E5%90%8E%E6%98%AF%E5%85%AC%E5%85%B1%E6%95%B0%E6%8D%AE%E5%BA%93%E6%8C%96%E6%8E%98"><span class="toc-number">5.1.0.3.</span> <span class="toc-text">最后是公共数据库挖掘</span></a></li></ol></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link"><span class="toc-number">5.2.</span> <span class="toc-text"></span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%E5%9C%A8%E5%8D%95%E7%BB%86%E8%83%9E%E5%A4%A9%E5%9C%B0%E6%8C%81%E7%BB%AD%E6%9B%B4%E6%96%B0"><span class="toc-number">6.</span> <span class="toc-text">学习笔记在单细胞天地持续更新</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2023/09/18/pigudog/pigudog1-palantir/" title="pigudog1-palantir"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg3.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="pigudog1-palantir"/></a><div class="content"><a class="title" href="/2023/09/18/pigudog/pigudog1-palantir/" title="pigudog1-palantir">pigudog1-palantir</a><time datetime="2023-09-17T16:00:00.000Z" title="Created 2023-09-18 00:00:00">2023-09-18</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/16/statistic/statistic7%20-%20Analysis%20of%20variance%20and%20regression/" title="statistic7 - Analysis of variance and regression"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic7 - Analysis of variance and regression"/></a><div class="content"><a class="title" href="/2023/02/16/statistic/statistic7%20-%20Analysis%20of%20variance%20and%20regression/" title="statistic7 - Analysis of variance and regression">statistic7 - Analysis of variance and regression</a><time datetime="2023-02-15T16:00:00.000Z" title="Created 2023-02-16 00:00:00">2023-02-16</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/15/statistic/statistic6%20-%20Hypothesis%20testing/" title="statistic6 - Hypothesis testing"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic6 - Hypothesis testing"/></a><div class="content"><a class="title" href="/2023/02/15/statistic/statistic6%20-%20Hypothesis%20testing/" title="statistic6 - Hypothesis testing">statistic6 - Hypothesis testing</a><time datetime="2023-02-14T16:00:00.000Z" title="Created 2023-02-15 00:00:00">2023-02-15</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/13/statistic/statistic4%20-%20Sampling/" title="statistic4 - Sampling"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic4 - Sampling"/></a><div class="content"><a class="title" href="/2023/02/13/statistic/statistic4%20-%20Sampling/" title="statistic4 - Sampling">statistic4 - Sampling</a><time datetime="2023-02-12T16:00:00.000Z" title="Created 2023-02-13 00:00:00">2023-02-13</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/02/11/statistic/statistic3%20-%20Law%20of%20large%20numbers%20and%20central%20limit%20law/" title="statistic0 - StatQuest"><img src="https://file.crazywong.com/gh/jerryc127/CDN@latest/cover/default_bg2.png0,limit_1/sharpen,100" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="statistic0 - StatQuest"/></a><div class="content"><a class="title" href="/2023/02/11/statistic/statistic3%20-%20Law%20of%20large%20numbers%20and%20central%20limit%20law/" title="statistic0 - StatQuest">statistic0 - StatQuest</a><time datetime="2023-02-10T16:00:00.000Z" title="Created 2023-02-11 00:00:00">2023-02-11</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2022 - 2023 By pigudog</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div><div class="footer_custom_text">Hi, welcome to my <a target="_blank" rel="noopener" href="https://pigudog.com">blog</a>!</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="Switch Between Traditional Chinese And Simplified Chinese">简</button><button id="darkmode" type="button" title="Switch Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><div class="js-pjax"></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>